{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "bda29970-525e-4c27-b168-335e48e2f4a4",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os \n",
    "import sys\n",
    "import pandas as pd \n",
    "import numpy as np\n",
    "import glob \n",
    "from Bio.Seq import Seq\n",
    "from Bio import SeqIO\n",
    "from scipy import stats\n",
    "import multiprocessing\n",
    "from multiprocessing import  Pool\n",
    "import time\n",
    "import logging\n",
    "from functools import partial\n",
    "import random\n",
    "from functools import reduce\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "86799b33-fc62-4f51-bab8-acdfce855d7c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "80"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "random.randint(0,100)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e8060e91-5ac7-421b-91f6-edb57f0e18ea",
   "metadata": {},
   "source": [
    "\"\"\"\n",
    "Workflow: \n",
    "\n",
    "1. index genomes \n",
    "    - smRNA (bowtie1)\n",
    "    - mRNA (genes, pseudogenes, lincRNAs, transposons...) (bowtie2)\n",
    "    - genome (remove non-chimeras) (bowtie2)\n",
    "    \n",
    "2. Collapse trim reads of adapters and collapse fastq files.\n",
    "\n",
    "3. Index collapsed fasta file \n",
    "\n",
    "4. Assign which chimeras have smRNAs. \n",
    "    - align smRNAs to collapsed fasta index. \n",
    "    - convert to bedfile.\n",
    "    - bowtie1 settings: -m 1 -v0\n",
    "    \n",
    "5. Parse smRNA from target RNA. \n",
    "\n",
    "6. Align target RNA to targets\n",
    "    - convert to bedfile\n",
    "    - inner join to table with \n",
    "    - bowtie2\n",
    "    \n",
    "7. Extend coordinates by 10 nt on each side of target, extract transcriptomic sequence. \n",
    "\n",
    "8. Output fasta file in the form >hybridN \\n TARGET&smRNA. Fold with RNAup.\n",
    "\n",
    "9. Merge RNAup output with table.\n",
    "\n",
    "\"\"\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "33a5a68f-8543-4ab6-ae1e-dbc3bed4a196",
   "metadata": {},
   "outputs": [],
   "source": [
    "def timing(f):\n",
    "    \"\"\"\n",
    "    Helper function for timing other functions\n",
    "    Parameters\n",
    "    ----------\n",
    "    f : function\n",
    "    Returns\n",
    "    -------\n",
    "    function\n",
    "        new function wrap with timer and logging \n",
    "    \"\"\"\n",
    "    def wrap(*args):\n",
    "        time1 = time.time()\n",
    "        ret = f(*args)\n",
    "        time2 = time.time()\n",
    "        logging.debug('{:s} function took {:.10f} s'.format(f.__name__, (time2-time1)))\n",
    "        return ret\n",
    "    \n",
    "    return wrap\n",
    "logging.basicConfig(level=logging.DEBUG)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "1e3b065c-5d62-4de8-9226-bcac9d8d4a65",
   "metadata": {},
   "outputs": [],
   "source": [
    "def parallelize_dataframe(df, func, n_cores = 20):\n",
    "    \n",
    "    df_split = np.array_split(df, n_cores)\n",
    "    pool = Pool(n_cores)\n",
    "    df = pd.concat(pool.map(func, df_split))\n",
    "    pool.close()\n",
    "    pool.join()\n",
    "\n",
    "    return df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "9a8ba733-6879-48e3-b8db-5cfac81866f3",
   "metadata": {},
   "outputs": [],
   "source": [
    "# ex of parallelization of dataframe\n",
    "#fin = parallelize_dataframe(df, ttest, n_cores = 30)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "07e6fa27-abde-4abd-ae6b-a09bcc00cdf5",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "def btindex(fasta, idx_path) : \n",
    "        \n",
    "    os.system(f\"sh ./bowtie_build.sh {fasta} {idx_path}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "39c8ef36-f63a-4df2-a754-19de19f02162",
   "metadata": {},
   "outputs": [],
   "source": [
    "def bt2index(fasta, idx_path) : \n",
    "    \n",
    "    os.system(f\"sh ./bowtie2_build.sh {fasta} {idx_path}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "e70b0559-3529-4328-b40a-babcd818d12e",
   "metadata": {},
   "outputs": [],
   "source": [
    "def trim_adapter(fastq, outdir, outname, adapter = \"AGATCGGAAGAGC\", min_len = 15) : \n",
    "    \n",
    "    \"\"\"\n",
    "    Inputs : \n",
    "        - output from collapse_reads() \n",
    "        - adapter sequence\n",
    "    \n",
    "    Outputs : \n",
    "        - trimmed fastq.gz file\n",
    "    \"\"\"\n",
    "\n",
    "    os.system(f\"sh ./trim_adapter.sh {adapter} {min_len} {fastq} {outdir}\")\n",
    "    \n",
    "    return outname\n",
    "    \n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "ae7a1ff5-4fd1-45eb-8b47-4410afabc61a",
   "metadata": {},
   "outputs": [],
   "source": [
    "def collapse_reads(fastq, collapsed) : \n",
    "    \n",
    "    \"\"\"\n",
    "    Inputs : \n",
    "        - gzipped fastq file (from trim_adapter)\n",
    "        \n",
    "    Outputs : \n",
    "        - collapsed fasta file in the form\n",
    "        - >sequence:count\n",
    "          sequence\n",
    "        \n",
    "        returns path to collapsed fasta\n",
    "    \"\"\"\n",
    "    \n",
    "    os.system(f\"sh ./collapse.sh {fastq} {collapsed}\")\n",
    "    \n",
    "    return collapsed"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "ec5764b4-8f03-4a4b-820d-8c435e84b4ef",
   "metadata": {},
   "outputs": [],
   "source": [
    "def align_genome(fasta, genome_index, outname) : \n",
    "    \n",
    "    \"\"\"\n",
    "    Inputs : \n",
    "        - trimmed & collapsed fasta file \n",
    "    \n",
    "    Outputs : \n",
    "        - fasta file of non-aligned reads\n",
    "    \"\"\"\n",
    "    \n",
    "    os.system(f\"sh ./bowtie_align_genome.sh {fasta} {genome_index} {outname}\")\n",
    "    \n",
    "    return outname"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "12f598c9-07ac-4ff3-9fff-3cbf652e89f9",
   "metadata": {},
   "outputs": [],
   "source": [
    "def align_smRNA_chimera(fasta, index, outname) : \n",
    "    \n",
    "    \"\"\"\n",
    "    Inputs :\n",
    "        - fasta file of smRNAs\n",
    "        - index of chimeras\n",
    "        - index name\n",
    "        - outdir\n",
    "    \n",
    "    Outputs :\n",
    "        - dataframe in bed format\n",
    "        \n",
    "    \"\"\"\n",
    "        \n",
    "    os.system(f\"sh ./bowtie_align_smRNA_chimera.sh {fasta} {index} {outname}\")\n",
    "    \n",
    "    return outname\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "19c219c5-eee0-4689-920a-dd676f65e208",
   "metadata": {},
   "outputs": [],
   "source": [
    "def align_targets_transcriptome(fasta, index, outname) : \n",
    "    \n",
    "    \"\"\"\n",
    "    Inputs :\n",
    "        - fasta file of target RNAs parsed from chimeras\n",
    "        - index name\n",
    "        - outdir\n",
    "    \n",
    "    Outputs :\n",
    "        - dataframe in bed format\n",
    "        \n",
    "    \"\"\"\n",
    "        \n",
    "    os.system(f\"sh ./bowtie2_align_targets.sh {fasta} {index} {outname}\")\n",
    "    \n",
    "    return outname\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "1886b76d-e3ba-41cd-8acc-274214118a76",
   "metadata": {},
   "outputs": [],
   "source": [
    "def parse_target(target, smRNA) : \n",
    "    \n",
    "    splt = target.split(smRNA)\n",
    "    \n",
    "    return max(splt, key = len)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "beac04cb-aaa2-4357-aebe-2046cab514b2",
   "metadata": {},
   "outputs": [],
   "source": [
    "def process_smRNA_alignment(dat) : \n",
    "    \n",
    "    # make sure smRNA aligns sense to reference\n",
    "    dat = dat.query('smRNA_strand == \"+\"').reset_index(drop = True)\n",
    "        \n",
    "    # split clash read into count and seq cols\n",
    "    dat[['chimera', 'count']] = dat.clash_read.str.split(\":\",expand=True)\n",
    "    \n",
    "    # remove smRNA sequence from chimera\n",
    "    dat['target_seq'] = dat.apply(lambda x : parse_target(x['chimera'], x['smRNA_seq']), axis = 1)    \n",
    "    \n",
    "    dat = dat.drop(columns = ['score'])\n",
    "    # return the dataframe\n",
    "    return dat\n",
    "    \n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "20c18aa3-8b39-47c0-831a-692b3e8bfe90",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_seq(records, chrom, start, end) : \n",
    "    \n",
    "    seq = records[chrom].seq\n",
    "    \n",
    "    return (str(seq)[start : end])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "5713a8f0-e121-445d-aac7-ac3c664da23d",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_target_sequence(df, *args) :\n",
    "    \n",
    "    records = args['records']\n",
    "    \n",
    "    df['pad_up'] = df.apply(lambda x : x['target_start']-10 if x['target_start']-10 > 0 else 0, axis = 1 )\n",
    "    \n",
    "    df['pad_down'] = df.apply(lambda x : x['target_end']+10, axis = 1 )\n",
    "    \n",
    "    df['target_pad'] = df.apply(lambda x : get_seq(records, x['target'], x['pad_up'], x['pad_down']) , axis = 1 )\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "3b1e21dc-899e-4aec-895a-8d072bdba800",
   "metadata": {},
   "outputs": [],
   "source": [
    "def RNAup(fasta) : \n",
    "    \n",
    "    outname = fasta.replace(\".fa\", \"\") + \".tsv\"\n",
    "    \n",
    "    os.system(f\"sh ./RNAup.sh {fasta} {outname}\")\n",
    "\n",
    "def parallel_RNAup(files, n_cores = 5) : \n",
    "    \n",
    "    pool = Pool(n_cores)\n",
    "    pool.map(RNAup, files)\n",
    "    pool.close()\n",
    "    pool.join()\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "4ba23ffa-12f5-4845-bbf2-abac3e5a47d4",
   "metadata": {},
   "outputs": [],
   "source": [
    "def run_RNAup(RNAup_input, outdir, ncores = 4) : \n",
    "    RNAup_out = RNAup_input.replace(\".fa\", \".tsv\")\n",
    "    rand_num = random.randint(0,10000)\n",
    "    if not os.path.exists(RNAup_out) :\n",
    "        print(\"Folding RNA\")\n",
    "        i = 0\n",
    "        with open(RNAup_input, 'r') as f : \n",
    "            for line in f : \n",
    "                i += 1\n",
    "        f.close() \n",
    "\n",
    "        Nlines = i / 2\n",
    "        chunksize = 10000\n",
    "        if Nlines > chunksize : \n",
    "            tmpdir = os.path.join(outdir, \"RNAup_tmp\")\n",
    "\n",
    "            if not os.path.exists(tmpdir) : \n",
    "                os.mkdir(tmpdir)\n",
    "\n",
    "            Nfiles = round(Nlines / chunksize,0)\n",
    "            os.system(f'split -l {chunksize} {RNAup_input} {tmpdir}/tmp{rand_num}')\n",
    "\n",
    "            files = glob.glob(f\"{tmpdir}/tmp{rand_num}*\")\n",
    "            parallel_RNAup(files, n_cores = ncores)\n",
    "\n",
    "            # join all RNAup files together\n",
    "            os.system(f\"cat {tmpdir}/tmp{rand_num}*.tsv > {RNAup_out}\")\n",
    "\n",
    "        else : \n",
    "            RNAup(RNAup_input)\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "56184359-4c63-439f-b726-4d5a1b390ec0",
   "metadata": {},
   "outputs": [],
   "source": [
    "def process_rna_up(RNAup_out) :\n",
    "    rna_up = pd.read_csv(RNAup_out, sep = \"\\t\", names = ['ID', 'info', 'interaction_seq'])\n",
    "    rna_up['ID'] = rna_up['ID'].str.replace(\">\", \"\")\n",
    "    rna_up['info_split'] = rna_up.apply(lambda x : [ i for i in x['info'].split(\" \") if i != \"\" ], axis = 1)\n",
    "    rna_up['pairing'] = rna_up['info_split'].str[0]\n",
    "    rna_up['dG'] = rna_up['info_split'].str[4]\n",
    "    rna_up['dG'] = rna_up['dG'].str.replace(\"(\", \"\", regex = False).astype(float)\n",
    "    rna_up['target_pairing_region']  = rna_up['info_split'].str[1]\n",
    "    rna_up['smRNA_pairing_region']  = rna_up['info_split'].str[3]\n",
    "\n",
    "    rna_up['target_tmp'] = rna_up['interaction_seq'].str.split(\"&\").str[0]\n",
    "    rna_up['len_target'] = rna_up['target_tmp'].str.len()\n",
    "\n",
    "    # remove target sites that are less than 15 nucleotides long\n",
    "    rna_up = rna_up.query('len_target >= 15').reset_index(drop = True)\n",
    "    rna_up = rna_up.drop(columns = ['target_tmp', 'len_target'])\n",
    "\n",
    "    # remove target sites with no predicted basepairing\n",
    "    rna_up['piRNA_pairing'] = rna_up['pairing'].str.split(\"&\", expand = False).str[1]\n",
    "    rna_up['N_piRNA_bp'] = rna_up.apply(lambda x : len( [ i for i in x['piRNA_pairing'] if i == \")\"]), axis = 1)\n",
    "\n",
    "    # fill in blanks in basepairing with \".\" (i.e. most of the time base pairing begins at position 2 of piRNA not 5' U)\n",
    "    # each piRNA starts at 1 and goes to 21\n",
    "    # if piRNA pairing region starts at 2 add 1 \".\" at beginning of string \".\"*start-1\n",
    "\n",
    "\n",
    "    rna_up['modified_smRNA_pairing'] = rna_up.apply(lambda x : \n",
    "                                                ( int(x['smRNA_pairing_region'].split(\",\")[0])-1 )*\".\" + x['piRNA_pairing'] + (21-int(x['smRNA_pairing_region'].split(\",\")[1]))*\".\", \n",
    "                                                axis = 1)\n",
    "\n",
    "    return rna_up"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 85,
   "id": "f0b29190-97f0-4025-93ea-7f9b22637969",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "def run(smRNA_fasta, target_fasta, genome_fasta, fastq, index_dir = os.getcwd(), outdir = os.getcwd(), ncores = 4) : \n",
    "    \n",
    "    ############################################################\n",
    "    # timing function\n",
    "    t1 = time.time()\n",
    "    \n",
    "    ############################################################\n",
    "    # check to make sure idx dir exists\n",
    "    if not os.path.exists(index_dir) : \n",
    "        os.mkdir(index_dir)\n",
    "    \n",
    "    ############################################################\n",
    "    # check to make sure out dir exists\n",
    "    if not os.path.exists(outdir) : \n",
    "        os.mkdir(outdir)\n",
    "        \n",
    "    ############################################################\n",
    "    # build smRNA_fasta (bt)\n",
    "    smRNA_name = os.path.basename(smRNA_fasta).replace(\".fa\", \"\")\n",
    "    if not os.path.exists( os.path.join( index_dir, f\"{smRNA_name}.1.ebwt\" ) ) :\n",
    "        print(\"Building smRNA index\")\n",
    "        smRNA_index = os.path.join( index_dir, f\"{smRNA_name}\" )\n",
    "        btindex(smRNA_fasta, smRNA_index)\n",
    "    else : \n",
    "        smRNA_index = os.path.join( index_dir, f\"{smRNA_name}\" )\n",
    "    \n",
    "    ############################################################\n",
    "    # build target_fasta index (bt2)\n",
    "    target_name = os.path.basename(target_fasta).replace(\".fa\", \"\")\n",
    "    if not os.path.exists( os.path.join( index_dir, f\"{target_name}.1.bt2\" ) ) :\n",
    "        print(\"Building target index\")\n",
    "        target_index = os.path.join( index_dir, f\"{target_name}\" )\n",
    "        bt2index(target_fasta, target_index)\n",
    "    else : \n",
    "        target_index = os.path.join( index_dir, f\"{target_name}\" )\n",
    "    \n",
    "    ############################################################\n",
    "    # build genome index (bt2)\n",
    "    genome_name = os.path.basename(genome_fasta).replace(\".fa\", \"\")\n",
    "    if not os.path.exists( os.path.join( index_dir, f\"{genome_name}.1.bt2\" ) ) :\n",
    "        print(\"Building genome index\")\n",
    "        genome_index = os.path.join( index_dir, f\"{genome_name}\" )\n",
    "        bt2index(genome_fasta, genome_index)\n",
    "    else : \n",
    "        genome_index = os.path.join( index_dir, f\"{genome_name}\" )\n",
    "    \n",
    "    ############################################################\n",
    "    # trim adapter\n",
    "    t_fq = os.path.join(outdir, os.path.basename(fastq).replace(\".fq.gz\", \"_trimmed.fq.gz\").replace(\".fastq.gz\", \"_trimmed.fq.gz\"))\n",
    "    if not os.path.exists(t_fq) : \n",
    "        print(\"Trimming adapters\")\n",
    "        t_fq = trim_adapter(fastq, outdir, t_fq)\n",
    "    \n",
    "    ############################################################\n",
    "    # collapse reads\n",
    "    c_fa = t_fq.replace(\".fq.gz\", \".uni.fa\")\n",
    "    print(c_fa)\n",
    "    if not os.path.exists(c_fa) : \n",
    "        print(\"Collapsing trimmed reads\")\n",
    "        c_fa = collapse_reads(t_fq, c_fa)\n",
    "    \n",
    "    ############################################################\n",
    "    # align collapsed reads to genome\n",
    "    genome_unmapped = c_fa.replace(\".fa\", \".genomeUnmapped.fa\")\n",
    "    if not os.path.exists(genome_unmapped) : \n",
    "        print(\"Aligning to genome\")\n",
    "        align_genome(c_fa, genome_index, genome_unmapped)\n",
    "    \n",
    "    ############################################################\n",
    "    # build bt index of unaligned reads\n",
    "    chimera_index_name = os.path.basename(genome_unmapped).replace(\".fa\", \"\")\n",
    "    if not os.path.exists( os.path.join( outdir, f\"{chimera_index_name}.1.ebwt\" ) ) :\n",
    "        print(\"Build chimera index\")\n",
    "        chimera_index = os.path.join( outdir, f\"{chimera_index_name}\" )\n",
    "        btindex(genome_unmapped, chimera_index)\n",
    "    else : \n",
    "        chimera_index = os.path.join( outdir, f\"{chimera_index_name}\" )\n",
    "    \n",
    "    ############################################################\n",
    "    # align smRNA fasta to chimera index\n",
    "    smRNA_aligned = genome_unmapped.replace(\".fa\", \".smRNAaligned\")\n",
    "    smRNA_aligned_bed = genome_unmapped.replace(\".fa\", \".smRNAaligned.bed\")\n",
    "    if not os.path.exists(smRNA_aligned_bed) : \n",
    "        print(\"Align smRNA to chimera index\")\n",
    "        align_smRNA_chimera(smRNA_fasta, chimera_index, smRNA_aligned)\n",
    "    \n",
    "    ############################################################\n",
    "    # parse smRNA alignment\n",
    "    print(\"Process smRNA alignment\")\n",
    "    dat = pd.read_csv(smRNA_aligned_bed, \n",
    "                      sep = \"\\t\",\n",
    "                      names = ['clash_read', 'smRNA_start', 'smRNA_end', 'smRNA', 'score', 'smRNA_strand', 'smRNA_seq'], \n",
    "                      header = 0)\n",
    "    master = parallelize_dataframe(dat, process_smRNA_alignment, n_cores = ncores)\n",
    "    master = master.query('target_seq.str.len() >= 15').reset_index(drop = True)\n",
    "    \n",
    "    # write fasta file with target sequences to align to the genome\n",
    "    lines = ''\n",
    "    for index,row in master.iterrows() : \n",
    "        lines += f\">{row['clash_read']}\\n{row['target_seq']}\\n\"\n",
    "        \n",
    "    targets_seq_parsed = f\"{smRNA_aligned}.targets.fa\"\n",
    "    op = open(targets_seq_parsed, 'w')\n",
    "    op.write(lines)\n",
    "    op.close() \n",
    "    lines = ''\n",
    "    \n",
    "    ############################################################\n",
    "    # align target fasta to targets\n",
    "    targets_aligned = targets_seq_parsed.replace(\".fa\", \".targetsAligned\")\n",
    "    targets_aligned_bed = targets_seq_parsed.replace(\".fa\", \".targetsAligned.bed\")\n",
    "    if not os.path.exists(targets_aligned_bed) : \n",
    "        print(\"Align targets to target RNA index\")\n",
    "        align_targets_transcriptome(targets_seq_parsed, target_index, targets_aligned)\n",
    "    \n",
    "    \n",
    "    ############################################################\n",
    "    # merge aligned targets to master table\n",
    "    print(\"Process mRNA alignment\")\n",
    "    dat = pd.read_csv(targets_aligned_bed, \n",
    "                      sep = \"\\t\",\n",
    "                      names = ['target', 'target_start', 'target_end', 'clash_read', 'score', 'target_strand', 'aligned_region'], \n",
    "                      header = 0)\n",
    "    \n",
    "    dat = dat.drop(columns = ['score'])\n",
    "    \n",
    "    ############################################################\n",
    "    # merge aligned targets to df\n",
    "    master = dat.merge(master, on = ['clash_read'], how = 'left')\n",
    "    \n",
    "    ############################################################\n",
    "    # extend target region by 0 nt on both sides\n",
    "    records = SeqIO.to_dict(SeqIO.parse(target_fasta, 'fasta'))\n",
    "    \n",
    "    pad = 5\n",
    "    master['pad_up'] = master.apply(lambda x : x['target_start']-pad if x['target_start']-pad > 0 else 0, axis = 1 )\n",
    "    master['pad_down'] = master.apply(lambda x : x['target_end']+pad, axis = 1 )\n",
    "    master['target_pad'] = master.apply(lambda x : get_seq(records, x['target'], x['pad_up'], x['pad_down']) , axis = 1 )\n",
    "    \n",
    "    master['ID'] = [ f'id_{i}' for i in range(0,master.shape[0]) ]\n",
    "    \n",
    "    ############################################################\n",
    "    # iterate through df_merge and make RNAup input \n",
    "    lines = ''\n",
    "    for index,row in master.iterrows() : \n",
    "        lines += f\">{row['ID']}\\n{row['target_pad']}&{row['smRNA_seq']}\\n\"\n",
    "        \n",
    "    RNAup_input = f\"{targets_aligned}.RNAup.fa\"\n",
    "    op = open(RNAup_input, 'w')\n",
    "    op.write(lines)\n",
    "    op.close() \n",
    "    lines = ''\n",
    "    \n",
    "    ############################################################\n",
    "    # RNAup folding \n",
    "    print(\"RNAup\")\n",
    "    RNAup_out = RNAup_input.replace(\".fa\", \".tsv\")\n",
    "    \n",
    "    # check size of file if lines > chunk size break into smaller files for RNA folding\n",
    "    # fasta file has two lines per entry so divide i / 2 to get number of computations\n",
    "    run_RNAup(RNAup_input, outdir)\n",
    "    \n",
    "    rna_up = process_rna_up(RNAup_out)\n",
    "    \n",
    "    no_bp = rna_up.query('N_piRNA_bp == 0').reset_index(drop = True)\n",
    "    \n",
    "    rna_up = rna_up.query('N_piRNA_bp >= 4').reset_index(drop = True)\n",
    "\n",
    "    rna_up = rna_up[['ID', 'interaction_seq', 'pairing', 'dG', 'modified_smRNA_pairing', 'target_pairing_region', 'smRNA_pairing_region']]\n",
    "\n",
    "    ####################\n",
    "    # Do a refined search , for target RNA interactions with no predicted base-pairs re-do folding with \n",
    "    # better sequence prediction of where targeting might occur\n",
    "    no_bp = no_bp.merge(master, on = ['ID'], how = 'left')\n",
    "    no_bp['s'] = no_bp['target_pairing_region'].str.split(\",\").str[0].astype(int) - 5\n",
    "    no_bp['e'] = no_bp['target_pairing_region'].str.split(\",\").str[1].astype(int) + 5\n",
    "    \n",
    "    def subseq(seq, s, e) : \n",
    "        return seq[int(s) : int(e)]    \n",
    "    \n",
    "    no_bp['refined_target_seq'] = no_bp.apply(lambda x : subseq(x['target_pad'], x['s']-1, x['e']), axis = 1)\n",
    "\n",
    "    no_bp = no_bp.query('refined_target_seq.str.len() >= 15')\n",
    "    \n",
    "    ####################\n",
    "    # adjust target start and end from master table \n",
    "    ids = no_bp['ID'].tolist()\n",
    "    master_sub = master.query('ID.isin(@ids)').reset_index(drop = True)\n",
    "\n",
    "    no_bp_sub = no_bp[['ID', 's', 'e', 'refined_target_seq']]\n",
    "\n",
    "    master_sub = master_sub.merge(no_bp_sub, on = ['ID'], how = 'left')\n",
    "    \n",
    "    lines = ''\n",
    "    for index,row in master_sub.iterrows() : \n",
    "        lines += f\">{row['ID']}\\n{row['refined_target_seq']}&{row['smRNA_seq']}\\n\"\n",
    "        \n",
    "    RNAup_input = f\"{targets_aligned}.RNAup.refined.fa\"\n",
    "    op = open(RNAup_input, 'w')\n",
    "    op.write(lines)\n",
    "    op.close() \n",
    "    lines = ''\n",
    "    \n",
    "    RNAup_out = RNAup_input.replace(\".fa\", \".tsv\")\n",
    "    run_RNAup(RNAup_input, outdir)\n",
    "    rna_up_refined = process_rna_up(RNAup_out)\n",
    "    rna_up_refined = rna_up_refined.query('N_piRNA_bp >= 4').reset_index(drop = True)\n",
    "    rna_up_refined = rna_up_refined[['ID', 'interaction_seq', 'pairing', 'dG', 'modified_smRNA_pairing', 'target_pairing_region', 'smRNA_pairing_region']]\n",
    "    \n",
    "    master_sub = master.query('ID.isin(@ids)').reset_index(drop = True)\n",
    "    \n",
    "    ####################\n",
    "    # merge rna_up to master\n",
    "    master_rnaup = rna_up.merge(master, on = ['ID'], how = 'left')\n",
    "    \n",
    "    ####################\n",
    "    # merge rna_up_refined to master table\n",
    "    master_sub_rnaup = rna_up_refined.merge(master_sub, on = ['ID'], how = 'left')\n",
    "    \n",
    "    ####################\n",
    "    # merge master tables\n",
    "    master_rnaup = pd.concat([master_rnaup, master_sub_rnaup], ignore_index = True)\n",
    "    master_rnaup = master_rnaup.rename(columns = {'count':'clash_count'})\n",
    "    master_rnaup['clash_count'] = master_rnaup['clash_count'].astype(float)\n",
    "    \n",
    "    # calculate transcriptomic coordinates of target site\n",
    "    master_rnaup['target_site_start'] = master_rnaup['target_start'] + master_rnaup['target_pairing_region'].str.split(\",\").str[0].astype(int)\n",
    "    master_rnaup['target_site_end'] = master_rnaup['target_start'] + master_rnaup['target_pairing_region'].str.split(\",\").str[1].astype(int)\n",
    "    master_rnaup['strand'] = '-'\n",
    "    \n",
    "    # clean up master table\n",
    "    master_rnaup = master_rnaup[['ID', 'target', 'target_site_start', 'target_site_end', 'smRNA', 'clash_count', 'strand', 'dG', 'modified_smRNA_pairing', 'interaction_seq', 'pairing', 'clash_read']]\n",
    "    \n",
    "    # sort values by count\n",
    "    master_rnaup = master_rnaup.sort_values(by = ['clash_count'], ascending = False).reset_index(drop = True)\n",
    "    print(f\"total target sites remaining == {master_rnaup.shape[0]}\")\n",
    "\n",
    "    # combine identical target sites, groupby pairing, target, smRNA, target_start, target_end, strand, \n",
    "    master_rnaup_grouped = master_rnaup.groupby( by = ['target', 'target_site_start', 'target_site_end', 'smRNA', 'modified_smRNA_pairing', 'strand', 'interaction_seq']).agg({'dG':'mean', 'clash_count':'sum'}).reset_index()\n",
    "    \n",
    "    t2 = time.time()\n",
    "    print(f\"CLASH Analysis Pipeline took {round( (t2-t1), 5)} seconds to run!\")\n",
    "    print(f\"total target sites remaining == {master_rnaup_grouped.shape[0]}\")\n",
    "    return([ master_rnaup, master_rnaup_grouped ])\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 86,
   "id": "e9068a50-3455-4dcf-adb1-25fe7f0c61f0",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/fs/ess/PAS1473/ben_past_projects/prg1_clash_rep1/prg1CLASH_1_trimmed.uni.fa\n",
      "Process smRNA alignment\n",
      "Process mRNA alignment\n",
      "RNAup\n",
      "total target sites remaining == 445507\n",
      "CLASH Analysis Pipeline took 178.84057 seconds to run!\n",
      "total target sites remaining == 303535\n"
     ]
    }
   ],
   "source": [
    "res = run(\n",
    "    smRNA_fasta = \"/fs/ess/PCON0160/ben/genomes/c_elegans/WS279/piRNA.fa\",\n",
    "    target_fasta = \"/fs/ess/PCON0160/ben/genomes/c_elegans/WS279/ce_ws279.linc.pseudo.pc.repbase.fa\",\n",
    "    genome_fasta = \"/fs/ess/PCON0160/ben/genomes/c_elegans/WS279/c_elegans.PRJNA13758.WS279.genomic.fa\",\n",
    "    fastq = \"/fs/ess/PAS1473/deep_sequencing/prg1CLASH/fastq/prg1CLASH_1.fastq.gz\",\n",
    "    index_dir = \"/fs/ess/PAS1473/ben_past_projects/clash_index\", \n",
    "    outdir = \"/fs/ess/PAS1473/ben_past_projects/prg1_clash_rep1\",\n",
    "    ncores = 16\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 87,
   "id": "35fa4fea-4cd2-4853-9fc0-d93f3cdb0676",
   "metadata": {},
   "outputs": [],
   "source": [
    "ungrouped = res[0]\n",
    "grouped = res[1]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 88,
   "id": "2a7cc283-8921-4cfe-a5ad-a9a09e41ffd7",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>target</th>\n",
       "      <th>target_site_start</th>\n",
       "      <th>target_site_end</th>\n",
       "      <th>smRNA</th>\n",
       "      <th>modified_smRNA_pairing</th>\n",
       "      <th>strand</th>\n",
       "      <th>interaction_seq</th>\n",
       "      <th>dG</th>\n",
       "      <th>clash_count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>303474</th>\n",
       "      <td>WBGene00306003,W04B5.8</td>\n",
       "      <td>2115.0</td>\n",
       "      <td>2135.0</td>\n",
       "      <td>21ur-161</td>\n",
       "      <td>.)))))))))).))..)))))</td>\n",
       "      <td>-</td>\n",
       "      <td>AUCUCGAGGCCCUGGAUGAGA&amp;UCUCAUCCGGUCCAAGAGGU</td>\n",
       "      <td>-21.81125</td>\n",
       "      <td>23.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                        target  target_site_start  target_site_end     smRNA  \\\n",
       "303474  WBGene00306003,W04B5.8             2115.0           2135.0  21ur-161   \n",
       "\n",
       "       modified_smRNA_pairing strand  \\\n",
       "303474  .)))))))))).))..)))))      -   \n",
       "\n",
       "                                   interaction_seq        dG  clash_count  \n",
       "303474  AUCUCGAGGCCCUGGAUGAGA&UCUCAUCCGGUCCAAGAGGU -21.81125         23.0  "
      ]
     },
     "execution_count": 88,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "grouped.query('smRNA == \"21ur-161\" and target == \"WBGene00306003,W04B5.8\"')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 89,
   "id": "1b804d82-fab5-46f1-ab94-5169c23aea1e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>target</th>\n",
       "      <th>target_site_start</th>\n",
       "      <th>target_site_end</th>\n",
       "      <th>smRNA</th>\n",
       "      <th>modified_smRNA_pairing</th>\n",
       "      <th>strand</th>\n",
       "      <th>interaction_seq</th>\n",
       "      <th>dG</th>\n",
       "      <th>clash_count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>248298</th>\n",
       "      <td>WBGene00019569,K09D9.12</td>\n",
       "      <td>1434.0</td>\n",
       "      <td>1452.0</td>\n",
       "      <td>21ur-4564</td>\n",
       "      <td>)))))))))))))))))....</td>\n",
       "      <td>-</td>\n",
       "      <td>GUGAUCUGCCAUCGUUCCA&amp;UGGAACGUGGCGGUCAC</td>\n",
       "      <td>-31.220000</td>\n",
       "      <td>2524.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>265823</th>\n",
       "      <td>WBGene00021015,W03G9.3,enu-3.3</td>\n",
       "      <td>810.0</td>\n",
       "      <td>827.0</td>\n",
       "      <td>21ur-2061</td>\n",
       "      <td>.)))))))....))))))...</td>\n",
       "      <td>-</td>\n",
       "      <td>GCCGCGAAUCCGUAUGUG&amp;CACAUACAAGGCGCGGC</td>\n",
       "      <td>-22.947500</td>\n",
       "      <td>2171.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>201848</th>\n",
       "      <td>WBGene00015405,C03H5.3</td>\n",
       "      <td>382.0</td>\n",
       "      <td>403.0</td>\n",
       "      <td>21ur-2634</td>\n",
       "      <td>)))))))..).))))))))))</td>\n",
       "      <td>-</td>\n",
       "      <td>AAUCGUGGCGCCAUCGCAAGAA&amp;UUCUUGCAAGUUGUCACGAUU</td>\n",
       "      <td>-22.625556</td>\n",
       "      <td>1884.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>117866</th>\n",
       "      <td>WBGene00008238,C50F4.16</td>\n",
       "      <td>994.0</td>\n",
       "      <td>1015.0</td>\n",
       "      <td>21ur-4204</td>\n",
       "      <td>))))).)))))))))))).))</td>\n",
       "      <td>-</td>\n",
       "      <td>CCUAAAAUUGCAUACACUCUUG&amp;UAAGAAUGUUGCAAUUUUCGG</td>\n",
       "      <td>-22.608500</td>\n",
       "      <td>1881.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>111411</th>\n",
       "      <td>WBGene00007709,C25A1.8,clec-87</td>\n",
       "      <td>510.0</td>\n",
       "      <td>526.0</td>\n",
       "      <td>21ur-2809</td>\n",
       "      <td>.))))))))))....))))..</td>\n",
       "      <td>-</td>\n",
       "      <td>GGACUUGUCAGAUUCAC&amp;GUGAAUCUGAAAAUGUCC</td>\n",
       "      <td>-25.623333</td>\n",
       "      <td>1561.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>148151</th>\n",
       "      <td>WBGene00010731,K10C3.4</td>\n",
       "      <td>4578.0</td>\n",
       "      <td>4597.0</td>\n",
       "      <td>21ur-1449</td>\n",
       "      <td>.)))))).)))....))))))</td>\n",
       "      <td>-</td>\n",
       "      <td>UUUCCCAUAAUUUGUGUGGC&amp;GCCACAACAGCUCAGGGAAA</td>\n",
       "      <td>-19.270000</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>148152</th>\n",
       "      <td>WBGene00010731,K10C3.4</td>\n",
       "      <td>4880.0</td>\n",
       "      <td>4899.0</td>\n",
       "      <td>21ur-3103</td>\n",
       "      <td>))))))))))))).)))))))</td>\n",
       "      <td>-</td>\n",
       "      <td>AUGUUCUUCCAGAGCAUUGA&amp;UCAAUGCUCUGGACAGAACGU</td>\n",
       "      <td>-22.060000</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>148153</th>\n",
       "      <td>WBGene00010731,K10C3.4</td>\n",
       "      <td>5034.0</td>\n",
       "      <td>5055.0</td>\n",
       "      <td>21ur-51</td>\n",
       "      <td>)))))))))))..))).))).</td>\n",
       "      <td>-</td>\n",
       "      <td>UCCCCGUCUCUCUCUCUAAUUA&amp;UAAUUAGAAGGCCCGGUGGA</td>\n",
       "      <td>-20.900000</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>148155</th>\n",
       "      <td>WBGene00010731,K10C3.4</td>\n",
       "      <td>5586.0</td>\n",
       "      <td>5602.0</td>\n",
       "      <td>21ur-1909</td>\n",
       "      <td>...))).)))..)).))))))</td>\n",
       "      <td>-</td>\n",
       "      <td>UUUUUUUGUUUAUCUCU&amp;GGAAAUAUUGCUAAAAAA</td>\n",
       "      <td>-7.330000</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>303534</th>\n",
       "      <td>WBGene00306081,T21G5.12</td>\n",
       "      <td>377.0</td>\n",
       "      <td>394.0</td>\n",
       "      <td>21ur-68</td>\n",
       "      <td>))))))))..)))))).....</td>\n",
       "      <td>-</td>\n",
       "      <td>ACCGGUUGCUGGAUCACA&amp;UGUGAUCUCUACCGGU</td>\n",
       "      <td>-21.980000</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>303535 rows × 9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                target  target_site_start  target_site_end  \\\n",
       "248298         WBGene00019569,K09D9.12             1434.0           1452.0   \n",
       "265823  WBGene00021015,W03G9.3,enu-3.3              810.0            827.0   \n",
       "201848          WBGene00015405,C03H5.3              382.0            403.0   \n",
       "117866         WBGene00008238,C50F4.16              994.0           1015.0   \n",
       "111411  WBGene00007709,C25A1.8,clec-87              510.0            526.0   \n",
       "...                                ...                ...              ...   \n",
       "148151          WBGene00010731,K10C3.4             4578.0           4597.0   \n",
       "148152          WBGene00010731,K10C3.4             4880.0           4899.0   \n",
       "148153          WBGene00010731,K10C3.4             5034.0           5055.0   \n",
       "148155          WBGene00010731,K10C3.4             5586.0           5602.0   \n",
       "303534         WBGene00306081,T21G5.12              377.0            394.0   \n",
       "\n",
       "            smRNA modified_smRNA_pairing strand  \\\n",
       "248298  21ur-4564  )))))))))))))))))....      -   \n",
       "265823  21ur-2061  .)))))))....))))))...      -   \n",
       "201848  21ur-2634  )))))))..).))))))))))      -   \n",
       "117866  21ur-4204  ))))).)))))))))))).))      -   \n",
       "111411  21ur-2809  .))))))))))....))))..      -   \n",
       "...           ...                    ...    ...   \n",
       "148151  21ur-1449  .)))))).)))....))))))      -   \n",
       "148152  21ur-3103  ))))))))))))).)))))))      -   \n",
       "148153    21ur-51  )))))))))))..))).))).      -   \n",
       "148155  21ur-1909  ...))).)))..)).))))))      -   \n",
       "303534    21ur-68  ))))))))..)))))).....      -   \n",
       "\n",
       "                                     interaction_seq         dG  clash_count  \n",
       "248298         GUGAUCUGCCAUCGUUCCA&UGGAACGUGGCGGUCAC -31.220000       2524.0  \n",
       "265823          GCCGCGAAUCCGUAUGUG&CACAUACAAGGCGCGGC -22.947500       2171.0  \n",
       "201848  AAUCGUGGCGCCAUCGCAAGAA&UUCUUGCAAGUUGUCACGAUU -22.625556       1884.0  \n",
       "117866  CCUAAAAUUGCAUACACUCUUG&UAAGAAUGUUGCAAUUUUCGG -22.608500       1881.0  \n",
       "111411          GGACUUGUCAGAUUCAC&GUGAAUCUGAAAAUGUCC -25.623333       1561.0  \n",
       "...                                              ...        ...          ...  \n",
       "148151     UUUCCCAUAAUUUGUGUGGC&GCCACAACAGCUCAGGGAAA -19.270000          1.0  \n",
       "148152    AUGUUCUUCCAGAGCAUUGA&UCAAUGCUCUGGACAGAACGU -22.060000          1.0  \n",
       "148153   UCCCCGUCUCUCUCUCUAAUUA&UAAUUAGAAGGCCCGGUGGA -20.900000          1.0  \n",
       "148155          UUUUUUUGUUUAUCUCU&GGAAAUAUUGCUAAAAAA  -7.330000          1.0  \n",
       "303534           ACCGGUUGCUGGAUCACA&UGUGAUCUCUACCGGU -21.980000          1.0  \n",
       "\n",
       "[303535 rows x 9 columns]"
      ]
     },
     "execution_count": 89,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "grouped.sort_values(by = 'clash_count', ascending = False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 90,
   "id": "25b0a7fb-ae2d-4cc5-b477-e409f11c5436",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>target</th>\n",
       "      <th>target_site_start</th>\n",
       "      <th>target_site_end</th>\n",
       "      <th>smRNA</th>\n",
       "      <th>modified_smRNA_pairing</th>\n",
       "      <th>strand</th>\n",
       "      <th>interaction_seq</th>\n",
       "      <th>dG</th>\n",
       "      <th>clash_count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>17000</th>\n",
       "      <td>WBGene00000883,Y75B12B.2,cyn-7</td>\n",
       "      <td>62.0</td>\n",
       "      <td>81.0</td>\n",
       "      <td>21ur-4864</td>\n",
       "      <td>.))))))..)))))))))...</td>\n",
       "      <td>-</td>\n",
       "      <td>GGACGUAUCGUUAUGGAGCU&amp;AGCUCCUGGAUAUGUCU</td>\n",
       "      <td>-20.660000</td>\n",
       "      <td>255.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>168162</th>\n",
       "      <td>WBGene00012228,W03H9.2</td>\n",
       "      <td>492.0</td>\n",
       "      <td>511.0</td>\n",
       "      <td>21ur-4864</td>\n",
       "      <td>.)))))))))..)).))))))</td>\n",
       "      <td>-</td>\n",
       "      <td>GCUGGAAAUCAUCGGGAGCU&amp;AGCUCCUGGAUAUGUCUAGC</td>\n",
       "      <td>-24.174286</td>\n",
       "      <td>146.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>261844</th>\n",
       "      <td>WBGene00020696,T22F3.3,pygl-1</td>\n",
       "      <td>1291.0</td>\n",
       "      <td>1307.0</td>\n",
       "      <td>21ur-4864</td>\n",
       "      <td>.))))))).))..))))))).</td>\n",
       "      <td>-</td>\n",
       "      <td>CUAGACAUUGAGGAGCU&amp;AGCUCCUGGAUAUGUCUAG</td>\n",
       "      <td>-26.856667</td>\n",
       "      <td>80.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17001</th>\n",
       "      <td>WBGene00000883,Y75B12B.2,cyn-7</td>\n",
       "      <td>63.0</td>\n",
       "      <td>82.0</td>\n",
       "      <td>21ur-4864</td>\n",
       "      <td>.))))))..)))))))))...</td>\n",
       "      <td>-</td>\n",
       "      <td>GGACGUAUCGUUAUGGAGCU&amp;AGCUCCUGGAUAUGUCU</td>\n",
       "      <td>-20.660000</td>\n",
       "      <td>57.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>121088</th>\n",
       "      <td>WBGene00008514,F02A9.4,mrpl-44</td>\n",
       "      <td>114.0</td>\n",
       "      <td>130.0</td>\n",
       "      <td>21ur-4864</td>\n",
       "      <td>.))))...))).)))))))..</td>\n",
       "      <td>-</td>\n",
       "      <td>UAGCUCCUGGAUAUGUC&amp;GACAGAUUCUUGGAGCUG</td>\n",
       "      <td>-20.920000</td>\n",
       "      <td>48.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>158933</th>\n",
       "      <td>WBGene00011531,T06D10.1,rsbp-1</td>\n",
       "      <td>2335.0</td>\n",
       "      <td>2354.0</td>\n",
       "      <td>21ur-4864</td>\n",
       "      <td>.))))))))))))))))....</td>\n",
       "      <td>-</td>\n",
       "      <td>GACAUUUCCGUUCAGGAGCU&amp;AGCUCCUGGAUAUGUC</td>\n",
       "      <td>-16.090000</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>159109</th>\n",
       "      <td>WBGene00011550,T06G6.11</td>\n",
       "      <td>841.0</td>\n",
       "      <td>857.0</td>\n",
       "      <td>21ur-4864</td>\n",
       "      <td>....))))))).)))).))))</td>\n",
       "      <td>-</td>\n",
       "      <td>GCUAUGCAUCUCCAGGA&amp;UCCUGGAUAUGUCUAGC</td>\n",
       "      <td>-23.880000</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>139135</th>\n",
       "      <td>WBGene00010013,F54B3.1</td>\n",
       "      <td>14860.0</td>\n",
       "      <td>14883.0</td>\n",
       "      <td>21ur-4864</td>\n",
       "      <td>))))).)))))))))))))).</td>\n",
       "      <td>-</td>\n",
       "      <td>CUAGGCAUCAUCAUUCAGCAGCUA&amp;UAGCUCCUGGAUAUGUCUAG</td>\n",
       "      <td>-23.950000</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>150846</th>\n",
       "      <td>WBGene00010964,MTCE.26,ctc-1</td>\n",
       "      <td>1323.0</td>\n",
       "      <td>1338.0</td>\n",
       "      <td>21ur-4864</td>\n",
       "      <td>))).))))).).)))).....</td>\n",
       "      <td>-</td>\n",
       "      <td>ACAUUUUGCAGGACUA&amp;UAGCUCCUGGAUAUGU</td>\n",
       "      <td>-15.520000</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>146403</th>\n",
       "      <td>WBGene00010621,K07A12.2,egg-6</td>\n",
       "      <td>2669.0</td>\n",
       "      <td>2687.0</td>\n",
       "      <td>21ur-4864</td>\n",
       "      <td>))))))).).).))).)))..</td>\n",
       "      <td>-</td>\n",
       "      <td>UUAGCGUUUGUUGGAGCUG&amp;UAGCUCCUGGAUAUGUCUAG</td>\n",
       "      <td>-15.700000</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>236 rows × 9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                target  target_site_start  target_site_end  \\\n",
       "17000   WBGene00000883,Y75B12B.2,cyn-7               62.0             81.0   \n",
       "168162          WBGene00012228,W03H9.2              492.0            511.0   \n",
       "261844   WBGene00020696,T22F3.3,pygl-1             1291.0           1307.0   \n",
       "17001   WBGene00000883,Y75B12B.2,cyn-7               63.0             82.0   \n",
       "121088  WBGene00008514,F02A9.4,mrpl-44              114.0            130.0   \n",
       "...                                ...                ...              ...   \n",
       "158933  WBGene00011531,T06D10.1,rsbp-1             2335.0           2354.0   \n",
       "159109         WBGene00011550,T06G6.11              841.0            857.0   \n",
       "139135          WBGene00010013,F54B3.1            14860.0          14883.0   \n",
       "150846    WBGene00010964,MTCE.26,ctc-1             1323.0           1338.0   \n",
       "146403   WBGene00010621,K07A12.2,egg-6             2669.0           2687.0   \n",
       "\n",
       "            smRNA modified_smRNA_pairing strand  \\\n",
       "17000   21ur-4864  .))))))..)))))))))...      -   \n",
       "168162  21ur-4864  .)))))))))..)).))))))      -   \n",
       "261844  21ur-4864  .))))))).))..))))))).      -   \n",
       "17001   21ur-4864  .))))))..)))))))))...      -   \n",
       "121088  21ur-4864  .))))...))).)))))))..      -   \n",
       "...           ...                    ...    ...   \n",
       "158933  21ur-4864  .))))))))))))))))....      -   \n",
       "159109  21ur-4864  ....))))))).)))).))))      -   \n",
       "139135  21ur-4864  ))))).)))))))))))))).      -   \n",
       "150846  21ur-4864  ))).))))).).)))).....      -   \n",
       "146403  21ur-4864  ))))))).).).))).)))..      -   \n",
       "\n",
       "                                      interaction_seq         dG  clash_count  \n",
       "17000          GGACGUAUCGUUAUGGAGCU&AGCUCCUGGAUAUGUCU -20.660000        255.0  \n",
       "168162      GCUGGAAAUCAUCGGGAGCU&AGCUCCUGGAUAUGUCUAGC -24.174286        146.0  \n",
       "261844          CUAGACAUUGAGGAGCU&AGCUCCUGGAUAUGUCUAG -26.856667         80.0  \n",
       "17001          GGACGUAUCGUUAUGGAGCU&AGCUCCUGGAUAUGUCU -20.660000         57.0  \n",
       "121088           UAGCUCCUGGAUAUGUC&GACAGAUUCUUGGAGCUG -20.920000         48.0  \n",
       "...                                               ...        ...          ...  \n",
       "158933          GACAUUUCCGUUCAGGAGCU&AGCUCCUGGAUAUGUC -16.090000          1.0  \n",
       "159109            GCUAUGCAUCUCCAGGA&UCCUGGAUAUGUCUAGC -23.880000          1.0  \n",
       "139135  CUAGGCAUCAUCAUUCAGCAGCUA&UAGCUCCUGGAUAUGUCUAG -23.950000          1.0  \n",
       "150846              ACAUUUUGCAGGACUA&UAGCUCCUGGAUAUGU -15.520000          1.0  \n",
       "146403       UUAGCGUUUGUUGGAGCUG&UAGCUCCUGGAUAUGUCUAG -15.700000          1.0  \n",
       "\n",
       "[236 rows x 9 columns]"
      ]
     },
     "execution_count": 90,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "grouped.sort_values(by = 'clash_count', ascending = False).query('smRNA == \"21ur-4864\"')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 91,
   "id": "0d1b1adc-0f4a-49ad-9682-fe0898a32ac2",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "303532"
      ]
     },
     "execution_count": 91,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "grouped.sort_values(by = 'clash_count', ascending = False).query('dG <= 0').shape[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 92,
   "id": "5f231cb1-66e1-4b79-9338-9c93a2da5837",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "158378"
      ]
     },
     "execution_count": 92,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "grouped.sort_values(by = 'clash_count', ascending = False).query('dG <= -15 & clash_count > 1').shape[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 93,
   "id": "0a293dd6-21a1-4224-91bb-fafdbf38dcac",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "90718"
      ]
     },
     "execution_count": 93,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "grouped.sort_values(by = 'clash_count', ascending = False).query('dG <= -20 & clash_count > 1').shape[0]"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3.9 (Conda 2022.05) [python/3.9-2022.05]",
   "language": "python",
   "name": "python39_202205"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
