{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "bda29970-525e-4c27-b168-335e48e2f4a4",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os \n",
    "import sys\n",
    "import pandas as pd \n",
    "import numpy as np\n",
    "import glob \n",
    "from Bio.Seq import Seq\n",
    "from Bio import SeqIO\n",
    "from scipy import stats\n",
    "import multiprocessing\n",
    "from multiprocessing import  Pool\n",
    "import time\n",
    "import logging\n",
    "from functools import partial\n",
    "import random\n",
    "from functools import reduce\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "86799b33-fc62-4f51-bab8-acdfce855d7c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "52"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "random.randint(0,100)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e8060e91-5ac7-421b-91f6-edb57f0e18ea",
   "metadata": {},
   "source": [
    "\"\"\"\n",
    "Workflow: \n",
    "\n",
    "1. index genomes \n",
    "    - smRNA (bowtie1)\n",
    "    - mRNA (genes, pseudogenes, lincRNAs, transposons...) (bowtie2)\n",
    "    - genome (remove non-chimeras) (bowtie2)\n",
    "    \n",
    "2. Collapse trim reads of adapters and collapse fastq files.\n",
    "\n",
    "3. Index collapsed fasta file \n",
    "\n",
    "4. Assign which chimeras have smRNAs. \n",
    "    - align smRNAs to collapsed fasta index. \n",
    "    - convert to bedfile.\n",
    "    - bowtie1 settings: -m 1 -v0\n",
    "    \n",
    "5. Parse smRNA from target RNA. \n",
    "\n",
    "6. Align target RNA to targets\n",
    "    - convert to bedfile\n",
    "    - inner join to table with \n",
    "    - bowtie2\n",
    "    \n",
    "7. Extend coordinates by 10 nt on each side of target, extract transcriptomic sequence. \n",
    "\n",
    "8. Output fasta file in the form >hybridN \\n TARGET&smRNA. Fold with RNAup.\n",
    "\n",
    "9. Merge RNAup output with table.\n",
    "\n",
    "\"\"\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "33a5a68f-8543-4ab6-ae1e-dbc3bed4a196",
   "metadata": {},
   "outputs": [],
   "source": [
    "def timing(f):\n",
    "    \"\"\"\n",
    "    Helper function for timing other functions\n",
    "    Parameters\n",
    "    ----------\n",
    "    f : function\n",
    "    Returns\n",
    "    -------\n",
    "    function\n",
    "        new function wrap with timer and logging \n",
    "    \"\"\"\n",
    "    def wrap(*args):\n",
    "        time1 = time.time()\n",
    "        ret = f(*args)\n",
    "        time2 = time.time()\n",
    "        logging.debug('{:s} function took {:.10f} s'.format(f.__name__, (time2-time1)))\n",
    "        return ret\n",
    "    \n",
    "    return wrap\n",
    "logging.basicConfig(level=logging.DEBUG)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "1e3b065c-5d62-4de8-9226-bcac9d8d4a65",
   "metadata": {},
   "outputs": [],
   "source": [
    "def parallelize_dataframe(df, func, n_cores = 20):\n",
    "    \n",
    "    \"\"\"\n",
    "    Function to split up pandas dataframe and process it in \n",
    "    pieces for parallelization\n",
    "    \n",
    "    Inputs : \n",
    "        - dataframe\n",
    "        - function\n",
    "    \n",
    "    Outputs : \n",
    "        - datafrme \n",
    "    \"\"\"\n",
    "    \n",
    "    df_split = np.array_split(df, n_cores)\n",
    "    pool = Pool(n_cores)\n",
    "    df = pd.concat(pool.map(func, df_split))\n",
    "    pool.close()\n",
    "    pool.join()\n",
    "\n",
    "    return df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "07e6fa27-abde-4abd-ae6b-a09bcc00cdf5",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "def btindex(fasta, idx_path) : \n",
    "    \n",
    "    \"\"\" Index genome using bowtie \"\"\"\n",
    "    \n",
    "    os.system(f\"sh ./bowtie_build.sh {fasta} {idx_path}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "39c8ef36-f63a-4df2-a754-19de19f02162",
   "metadata": {},
   "outputs": [],
   "source": [
    "def bt2index(fasta, idx_path) : \n",
    "    \n",
    "    \"\"\" Index genome using bowtie2 \"\"\"\n",
    "    \n",
    "    os.system(f\"sh ./bowtie2_build.sh {fasta} {idx_path}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "e70b0559-3529-4328-b40a-babcd818d12e",
   "metadata": {},
   "outputs": [],
   "source": [
    "def trim_adapter(fastq, outdir, outname, adapter = \"AGATCGGAAGAGC\", min_len = 15) : \n",
    "    \n",
    "    \"\"\"\n",
    "    Inputs : \n",
    "        - output from collapse_reads() \n",
    "        - adapter sequence\n",
    "    \n",
    "    Outputs : \n",
    "        - trimmed fastq.gz file\n",
    "    \"\"\"\n",
    "\n",
    "    os.system(f\"sh ./trim_adapter.sh {adapter} {min_len} {fastq} {outdir}\")\n",
    "    \n",
    "    return outname\n",
    "    \n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "ae7a1ff5-4fd1-45eb-8b47-4410afabc61a",
   "metadata": {},
   "outputs": [],
   "source": [
    "def collapse_reads(fastq, collapsed) : \n",
    "    \n",
    "    \"\"\"\n",
    "    Inputs : \n",
    "        - gzipped fastq file (from trim_adapter)\n",
    "        \n",
    "    Outputs : \n",
    "        - collapsed fasta file in the form\n",
    "        - >sequence:count\n",
    "          sequence\n",
    "        \n",
    "        returns path to collapsed fasta\n",
    "    \"\"\"\n",
    "    \n",
    "    os.system(f\"sh ./collapse.sh {fastq} {collapsed}\")\n",
    "    \n",
    "    return collapsed"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "ec5764b4-8f03-4a4b-820d-8c435e84b4ef",
   "metadata": {},
   "outputs": [],
   "source": [
    "def align_genome(fasta, genome_index, outname) : \n",
    "    \n",
    "    \"\"\"\n",
    "    Inputs : \n",
    "        - trimmed & collapsed fasta file \n",
    "    \n",
    "    Outputs : \n",
    "        - fasta file of non-aligned reads\n",
    "    \"\"\"\n",
    "    \n",
    "    os.system(f\"sh ./bowtie_align_genome.sh {fasta} {genome_index} {outname}\")\n",
    "    \n",
    "    return outname"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "12f598c9-07ac-4ff3-9fff-3cbf652e89f9",
   "metadata": {},
   "outputs": [],
   "source": [
    "def align_smRNA_chimera(fasta, index, outname) : \n",
    "    \n",
    "    \"\"\"\n",
    "    Inputs :\n",
    "        - fasta file of smRNAs\n",
    "        - index of chimeras\n",
    "        - index name\n",
    "        - outdir\n",
    "    \n",
    "    Outputs :\n",
    "        - dataframe in bed format\n",
    "        \n",
    "    \"\"\"\n",
    "        \n",
    "    os.system(f\"sh ./bowtie1_align.sh {fasta} {index} {outname}\")\n",
    "    \n",
    "    return outname\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "803004b4-a213-453f-b5ef-349fce2044b6",
   "metadata": {},
   "outputs": [],
   "source": [
    "def bowtie1_align(fasta, index, outname, v = 0, m = 1) : \n",
    "    \n",
    "    \"\"\"\n",
    "    Inputs :\n",
    "        - fasta file of smRNAs\n",
    "        - index of chimeras\n",
    "        - index name\n",
    "        - outdir\n",
    "    \n",
    "    Outputs :\n",
    "        - dataframe in bed format\n",
    "        \n",
    "    \"\"\"\n",
    "        \n",
    "    os.system(f\"sh ./bowtie1_align.sh {fasta} {index} {outname}\")\n",
    "    \n",
    "    return outname"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "19c219c5-eee0-4689-920a-dd676f65e208",
   "metadata": {},
   "outputs": [],
   "source": [
    "def align_targets_transcriptome(fasta, index, outname) : \n",
    "    \n",
    "    \"\"\"\n",
    "    Inputs :\n",
    "        - fasta file of target RNAs parsed from chimeras\n",
    "        - index name\n",
    "        - outdir\n",
    "    \n",
    "    Outputs :\n",
    "        - dataframe in bed format\n",
    "        \n",
    "    \"\"\"\n",
    "        \n",
    "    os.system(f\"sh ./bowtie2_align_targets.sh {fasta} {index} {outname}\")\n",
    "    \n",
    "    return outname\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "1886b76d-e3ba-41cd-8acc-274214118a76",
   "metadata": {},
   "outputs": [],
   "source": [
    "def parse_target(target, smRNA) : \n",
    "    \n",
    "    \"\"\" Parse target sequence from chimera, take the longest string in the list after splitting \"\"\"\n",
    "    \n",
    "    splt = target.split(smRNA)\n",
    "    \n",
    "    return max(splt, key = len)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "beac04cb-aaa2-4357-aebe-2046cab514b2",
   "metadata": {},
   "outputs": [],
   "source": [
    "def process_smRNA_alignment(dat) : \n",
    "    \n",
    "    \"\"\"\n",
    "    Process smRNA alignment, split target RNA from smRNA\n",
    "    \n",
    "    \"\"\"\n",
    "    \n",
    "    # make sure smRNA aligns sense to reference\n",
    "    dat = dat.query('smRNA_strand == \"+\"').reset_index(drop = True)\n",
    "        \n",
    "    # split clash read into count and seq cols\n",
    "    dat[['chimera', 'count']] = dat.clash_read.str.split(\":\",expand=True)\n",
    "    \n",
    "    # remove smRNA sequence from chimera\n",
    "    dat['target_seq'] = dat.apply(lambda x : parse_target(x['chimera'], x['smRNA_seq']), axis = 1)    \n",
    "    \n",
    "    dat = dat.drop(columns = ['score'])\n",
    "    # return the dataframe\n",
    "    return dat\n",
    "    \n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "20c18aa3-8b39-47c0-831a-692b3e8bfe90",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_seq(records, chrom, start, end) : \n",
    "    \n",
    "    \"\"\" Returns genomic sequence on chrom from start to end \"\"\"\n",
    "    \n",
    "    seq = records[chrom].seq\n",
    "    \n",
    "    return (str(seq)[start : end])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "5713a8f0-e121-445d-aac7-ac3c664da23d",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_target_sequence(df, *args) :\n",
    "    \n",
    "    \"\"\" Add genomic target sequence to dataframe \"\"\"\n",
    "    \n",
    "    records = args['records']\n",
    "    \n",
    "    df['pad_up'] = df.apply(lambda x : x['target_start']-10 if x['target_start']-10 > 0 else 0, axis = 1 )\n",
    "    \n",
    "    df['pad_down'] = df.apply(lambda x : x['target_end']+10, axis = 1 )\n",
    "    \n",
    "    df['target_pad'] = df.apply(lambda x : get_seq(records, x['target'], x['pad_up'], x['pad_down']) , axis = 1 )\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "3b1e21dc-899e-4aec-895a-8d072bdba800",
   "metadata": {},
   "outputs": [],
   "source": [
    "def RNAup(fasta) : \n",
    "    \n",
    "    \"\"\" Run shell script for RNAup folding \"\"\"\n",
    "    \n",
    "    outname = fasta.replace(\".fa\", \"\") + \".tsv\"\n",
    "    \n",
    "    os.system(f\"sh ./RNAup.sh {fasta} {outname}\")\n",
    "\n",
    "def parallel_RNAup(files, n_cores = 5) : \n",
    "    \n",
    "    \"\"\" Parallelize RNAup folding function \"\"\"\n",
    "    \n",
    "    pool = Pool(n_cores)\n",
    "    pool.map(RNAup, files)\n",
    "    pool.close()\n",
    "    pool.join()\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "4ba23ffa-12f5-4845-bbf2-abac3e5a47d4",
   "metadata": {},
   "outputs": [],
   "source": [
    "def run_RNAup(RNAup_input, outdir, ncores = 4) : \n",
    "    \n",
    "    \"\"\" Run RNAup folding, if Nlines of fasta are greater than 100000 split up the file and use parallel function \"\"\"\n",
    "    \n",
    "    RNAup_out = RNAup_input.replace(\".fa\", \".tsv\")\n",
    "    rand_num = random.randint(0,10000)\n",
    "    if not os.path.exists(RNAup_out) :\n",
    "        print(\"Folding RNA\")\n",
    "        i = 0\n",
    "        with open(RNAup_input, 'r') as f : \n",
    "            for line in f : \n",
    "                i += 1\n",
    "        f.close() \n",
    "\n",
    "        Nlines = i / 2\n",
    "        chunksize = 10000\n",
    "        if Nlines > chunksize : \n",
    "            tmpdir = os.path.join(outdir, \"RNAup_tmp\")\n",
    "\n",
    "            if not os.path.exists(tmpdir) : \n",
    "                os.mkdir(tmpdir)\n",
    "\n",
    "            Nfiles = round(Nlines / chunksize,0)\n",
    "            os.system(f'split -l {chunksize} {RNAup_input} {tmpdir}/tmp{rand_num}')\n",
    "\n",
    "            files = glob.glob(f\"{tmpdir}/tmp{rand_num}*\")\n",
    "            parallel_RNAup(files, n_cores = ncores)\n",
    "\n",
    "            # join all RNAup files together\n",
    "            os.system(f\"cat {tmpdir}/tmp{rand_num}*.tsv > {RNAup_out}\")\n",
    "\n",
    "        else : \n",
    "            RNAup(RNAup_input)\n",
    "    "
   ]
  },
  {
   "cell_type": "raw",
   "id": "71651718-2ab0-4e27-964a-9c6cdfa2a7da",
   "metadata": {},
   "source": [
    ">id_9988\n",
    "((((..(((.((((((&))))))...)))))))   1,16  :   5,20  (-15.17 = -17.72 + 2.55)\n",
    "UGCACUGUUCGGGAAG&CUUCCCAUUAACUGCA\n",
    ">y\n",
    "((((((...(((((((&))))..))).))))))   5,20  :   1,16  (-14.60 = -18.75 + 4.14)\n",
    "CUUCCCAUUAACUGCA&UGCACUGUUCGGGAAG"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "86ef7285-5d2e-4a0c-b0eb-6dc79bccfce8",
   "metadata": {},
   "outputs": [],
   "source": [
    "def fix_pairing(s, smRNA, interaction, pairing, left_coords, right_coords) : \n",
    "    \n",
    "    \"\"\"\n",
    "    For some reason, after RNAup folding the position of the smRNA and target get switched.\n",
    "    This function finds if the RNAup entry1 or entry2 matches smRNA. entry1&entry2\n",
    "    If smRNA == entry1 must flip around the interaction seq and the pairing pattern.\n",
    "    \"\"\"\n",
    "    \n",
    "    if s.replace(\"U\", \"T\") in smRNA : \n",
    "        # if s in smRNA it means smRNA seq and target set order is switched, have to fix the interaction seq\n",
    "        # and base pairing\n",
    "        interaction_fixed = f\"\"\"{interaction.split(\"&\")[1]}&{interaction.split(\"&\")[0]}\"\"\"\n",
    "        pairing_fixed = f\"\"\"{pairing.split(\"&\")[1].replace(\")\", \"(\")}&{pairing.split(\"&\")[0].replace(\"(\", \")\")}\"\"\"\n",
    "        \n",
    "        return (interaction_fixed, pairing_fixed, right_coords, left_coords)\n",
    "    \n",
    "    else : \n",
    "        return (interaction, pairing, left_coords, right_coords)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "56184359-4c63-439f-b726-4d5a1b390ec0",
   "metadata": {},
   "outputs": [],
   "source": [
    "def process_rna_up(RNAup_out) :\n",
    "    \n",
    "    \n",
    "    \"\"\" Process RNAup output \"\"\"\n",
    "    \n",
    "    rna_up = pd.read_csv(RNAup_out, sep = \"\\t\", names = ['ID_info', 'info', 'interaction_seq'])\n",
    "    \n",
    "    rna_up['ID'] = rna_up['ID_info'].str.replace(\">\", \"\").str.split(\":\").str[0]\n",
    "    rna_up['smRNA_seq'] = rna_up['ID_info'].str.split(\":\").str[1]\n",
    "    rna_up = rna_up.drop(columns = ['ID_info'])\n",
    "    \n",
    "    rna_up['info_split'] = rna_up.apply(lambda x : [ i for i in x['info'].split(\" \") if i != \"\" ], axis = 1)\n",
    "    rna_up['pairing'] = rna_up['info_split'].str[0]\n",
    "    rna_up['dG'] = rna_up['info_split'].str[4]\n",
    "    rna_up['dG'] = rna_up['dG'].str.replace(\"(\", \"\", regex = False).astype(float)\n",
    "\n",
    "    rna_up['left_coord']  = rna_up['info_split'].str[1]\n",
    "    rna_up['right_coord']  = rna_up['info_split'].str[3]\n",
    "    \n",
    "    rna_up['tmp'] = rna_up['interaction_seq'].str.split(\"&\").str[0]\n",
    "\n",
    "    # check to see if tmp is in smRNA seq, if so have to reverse the pairing and interaction seq\n",
    "    rna_up['fix'] = rna_up.apply(\n",
    "        lambda x :\n",
    "        fix_pairing(x['tmp'], x['smRNA_seq'], x['interaction_seq'], \n",
    "                    x['pairing'], x['left_coord'], x['right_coord']), axis = 1)\n",
    "    \n",
    "    rna_up = rna_up.drop(columns = ['pairing', 'interaction_seq', 'tmp'])\n",
    "    rna_up[['interaction_seq', 'pairing', 'target_pairing_region', 'smRNA_pairing_region']] = pd.DataFrame(rna_up.fix.tolist(), index= rna_up.index)\n",
    "    rna_up['target'] = rna_up['interaction_seq'].str.split(\"&\").str[0]\n",
    "\n",
    "    # remove target sites that are less than 15 nucleotides long\n",
    "    rna_up = rna_up.query('target.str.len() >= 15').reset_index(drop = True)\n",
    "\n",
    "    # remove target sites with no predicted basepairing\n",
    "    rna_up['piRNA_pairing'] = rna_up['pairing'].str.split(\"&\", expand = False).str[1]\n",
    "    \n",
    "    rna_up['N_piRNA_bp'] = rna_up.apply(lambda x : len( [ i for i in x['piRNA_pairing'] if i == \")\"]), axis = 1)\n",
    "\n",
    "    # fill in blanks in basepairing with \".\" (i.e. most of the time base pairing begins at position 2 of piRNA not 5' U)\n",
    "    # each piRNA starts at 1 and goes to 21\n",
    "    # if piRNA pairing region starts at 2 add 1 \".\" at beginning of string \".\"*start-1\n",
    "\n",
    "\n",
    "    rna_up['modified_smRNA_pairing'] = rna_up.apply(lambda x : \n",
    "                                                ( int(x['smRNA_pairing_region'].split(\",\")[0])-1 )*\".\" + x['piRNA_pairing'] + (21-int(x['smRNA_pairing_region'].split(\",\")[1]))*\".\", \n",
    "                                                axis = 1)\n",
    "    \n",
    "    return rna_up"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 380,
   "id": "f123acac-0124-4286-935b-ed8cab779c26",
   "metadata": {},
   "outputs": [],
   "source": [
    "def collapse(data) : \n",
    "    \n",
    "    data = data.sort_values(by = ['target_start'], ascending = True)\n",
    "    data['target_smRNA'] = data.apply(lambda x : (x['target'], x['smRNA']), axis = 1)\n",
    "    uni = list(set(data['target_smRNA'].tolist()))\n",
    "    \n",
    "    k = 0\n",
    "    \n",
    "    for e,j in enumerate(uni) :\n",
    "        \n",
    "        target = j[0]\n",
    "        smRNA = j[1]\n",
    "        \n",
    "        df = data.query('target == @target and smRNA == @smRNA').reset_index()\n",
    "        idx = [k]\n",
    "        if df.shape[0] > 1 : \n",
    "            for i in range(0, df.shape[0]-1) :\n",
    "\n",
    "                # check next entry\n",
    "                s1 = int(df.iloc[i]['target_start'])\n",
    "                e1 = int(df.iloc[i]['target_end'])\n",
    "                l1 = e1 - s1 + 1\n",
    "\n",
    "                s2 = int(df.iloc[i+1]['target_start'])\n",
    "                e2 = int(df.iloc[i+1]['target_end'])\n",
    "                l2 = e2 - s2 + 1\n",
    "\n",
    "                overlap = len(range(max(s1, s2), min(e1, e2)+1))\n",
    "                o1 = overlap / l1\n",
    "                o2 = overlap / l2\n",
    "                if o1 > 0.9 and o2 > 0.9 : \n",
    "                    idx.append(k)\n",
    "                else : \n",
    "                    k += 1 \n",
    "                    idx.append(k)\n",
    "\n",
    "            df['group'] = idx\n",
    "        else :\n",
    "            df['group'] = k\n",
    "            \n",
    "        if e == 0 : \n",
    "            res = df\n",
    "        else : \n",
    "            res = pd.concat([res, df])\n",
    "        \n",
    "        k += 1\n",
    "        \n",
    "        print(f'{e} of {len(uni)} target smRNA pairs processed. ({ round(( (e)/(len(uni)) )*100,2) }%)', end='\\r')\n",
    "        \n",
    "        grped = res.groupby(['target', 'smRNA', 'group'], as_index = False).agg(\n",
    "                clash_count = ('clash_count', 'sum'), \n",
    "                dG = ('dG', 'min'),\n",
    "                start = ('target_start', 'min'), \n",
    "                end = ('target_end', 'max'),\n",
    "                pairing = ('modified_smRNA_pairing', 'unique'),\n",
    "                interaction = ('interaction_seq', 'unique')    \n",
    "            )\n",
    "        grped['strand'] = '+'\n",
    "        reformat = grped[['target', 'start', 'end', 'smRNA', 'clash_count', 'strand', 'dG', 'interaction', 'pairing']]\n",
    "        \n",
    "    return reformat"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 381,
   "id": "f0b29190-97f0-4025-93ea-7f9b22637969",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "def run(smRNA_fasta, target_fasta, genome_fasta, fastq, index_dir = os.getcwd(), outdir = os.getcwd(), ncores = 4) : \n",
    "    \n",
    "    ############################################################\n",
    "    # timing function\n",
    "    t1 = time.time()\n",
    "    \n",
    "    ############################################################\n",
    "    # check to make sure idx dir exists\n",
    "    if not os.path.exists(index_dir) : \n",
    "        os.mkdir(index_dir)\n",
    "    \n",
    "    ############################################################\n",
    "    # check to make sure out dir exists\n",
    "    if not os.path.exists(outdir) : \n",
    "        os.mkdir(outdir)\n",
    "        \n",
    "    ############################################################\n",
    "    # build smRNA_fasta (bt)\n",
    "    smRNA_name = os.path.basename(smRNA_fasta).replace(\".fa\", \"\")\n",
    "    smRNA_index = os.path.join( index_dir, f\"{smRNA_name}\" )\n",
    "    if not os.path.exists( os.path.join( index_dir, f\"{smRNA_name}.1.ebwt\" ) ) :\n",
    "        print(\"Building smRNA index\")\n",
    "        btindex(smRNA_fasta, smRNA_index)\n",
    "    \n",
    "    ############################################################\n",
    "    # build target_fasta index (bt2)\n",
    "    target_name = os.path.basename(target_fasta).replace(\".fa\", \".bowtie2\")\n",
    "    target_index = os.path.join( index_dir, f\"{target_name}\" )\n",
    "    if not os.path.exists( os.path.join( index_dir, f\"{target_name}.1.bt2\" ) ) :\n",
    "        print(\"Building target index\")\n",
    "        bt2index(target_fasta, target_index)\n",
    "        \n",
    "    ############################################################\n",
    "    # build target_fasta index (bt)\n",
    "    target_name = os.path.basename(target_fasta).replace(\".fa\", \".bowtie1\")\n",
    "    target_index_bowtie1 = os.path.join( index_dir, target_name )\n",
    "    if not os.path.exists( f\"{target_index_bowtie1}.1.ebwt\" ) :\n",
    "        print(\"Building target index\")\n",
    "        btindex(target_fasta, target_index_bowtie1)\n",
    "   \n",
    "    ############################################################\n",
    "    # build genome index (bt2)\n",
    "    genome_name = os.path.basename(genome_fasta).replace(\".fa\", \"\")\n",
    "    genome_index = os.path.join( index_dir, f\"{genome_name}\" )\n",
    "    if not os.path.exists( os.path.join( index_dir, f\"{genome_name}.1.bt2\" ) ) :\n",
    "        print(\"Building genome index\")\n",
    "        bt2index(genome_fasta, genome_index)\n",
    "    \n",
    "    ############################################################\n",
    "    # trim adapter\n",
    "    t_fq = os.path.join(outdir, os.path.basename(fastq).replace(\".fq.gz\", \"_trimmed.fq.gz\").replace(\".fastq.gz\", \"_trimmed.fq.gz\"))\n",
    "    if not os.path.exists(t_fq) : \n",
    "        print(\"Trimming adapters\")\n",
    "        t_fq = trim_adapter(fastq, outdir, t_fq)\n",
    "    \n",
    "    ############################################################\n",
    "    # collapse reads\n",
    "    c_fa = t_fq.replace(\".fq.gz\", \".uni.fa\")\n",
    "    print(c_fa)\n",
    "    if not os.path.exists(c_fa) : \n",
    "        print(\"Collapsing trimmed reads\")\n",
    "        c_fa = collapse_reads(t_fq, c_fa)\n",
    "    \n",
    "    ############################################################\n",
    "    # align collapsed reads to genome\n",
    "    genome_unmapped = c_fa.replace(\".fa\", \".genomeUnmapped.fa\")\n",
    "    if not os.path.exists(genome_unmapped) : \n",
    "        print(\"Aligning to genome\")\n",
    "        align_genome(c_fa, genome_index, genome_unmapped)\n",
    "    \n",
    "    ############################################################\n",
    "    # build bt index of unaligned reads\n",
    "    chimera_index_name = os.path.basename(genome_unmapped).replace(\".fa\", \"\")\n",
    "    chimera_index = os.path.join( outdir, f\"{chimera_index_name}\" )\n",
    "    if not os.path.exists( os.path.join( outdir, f\"{chimera_index_name}.1.ebwt\" ) ) :\n",
    "        print(\"Build chimera index\")\n",
    "        btindex(genome_unmapped, chimera_index)\n",
    "    \n",
    "    ############################################################\n",
    "    # align smRNA fasta to chimera index\n",
    "    smRNA_aligned = genome_unmapped.replace(\".fa\", \".smRNAaligned\")\n",
    "    smRNA_aligned_bed = genome_unmapped.replace(\".fa\", \".smRNAaligned.bed\")\n",
    "    if not os.path.exists(smRNA_aligned_bed) : \n",
    "        print(\"Align smRNA to chimera index\")\n",
    "        align_smRNA_chimera(smRNA_fasta, chimera_index, smRNA_aligned)\n",
    "    \n",
    "    ############################################################\n",
    "    # parse smRNA alignment\n",
    "    print(\"Process smRNA alignment\")\n",
    "    dat = pd.read_csv(smRNA_aligned_bed, \n",
    "                      sep = \"\\t\",\n",
    "                      names = ['clash_read', 'smRNA_start', 'smRNA_end', 'smRNA', 'score', 'smRNA_strand', 'smRNA_seq'], \n",
    "                      header = 0)\n",
    "    master = parallelize_dataframe(dat, process_smRNA_alignment, n_cores = ncores)\n",
    "    master = master.query('target_seq.str.len() >= 15').reset_index(drop = True)\n",
    "    \n",
    "    # write fasta file with target sequences to align to the genome\n",
    "    lines = ''\n",
    "    for index,row in master.iterrows() : \n",
    "        lines += f\">{row['clash_read']}\\n{row['target_seq']}\\n\"\n",
    "        \n",
    "    targets_seq_parsed = f\"{smRNA_aligned}.targets.fa\"\n",
    "    op = open(targets_seq_parsed, 'w')\n",
    "    op.write(lines)\n",
    "    op.close() \n",
    "    lines = ''\n",
    "    \n",
    "    ############################################################\n",
    "    # align target fasta to targets\n",
    "    targets_aligned = targets_seq_parsed.replace(\".fa\", \".targetsAligned\")\n",
    "    targets_aligned_bed = targets_seq_parsed.replace(\".fa\", \".targetsAligned.bed\")\n",
    "    if not os.path.exists(targets_aligned_bed) : \n",
    "        print(\"Align targets to target RNA index\")\n",
    "        align_targets_transcriptome(targets_seq_parsed, target_index, targets_aligned)\n",
    "    \n",
    "    \n",
    "    ############################################################\n",
    "    # merge aligned targets to master table\n",
    "    print(\"Process mRNA alignment\")\n",
    "    dat = pd.read_csv(targets_aligned_bed, \n",
    "                      sep = \"\\t\",\n",
    "                      names = ['target', 'target_start', 'target_end', 'clash_read', 'score', 'target_strand', 'aligned_region'], \n",
    "                      header = 0)\n",
    "    \n",
    "    dat = dat.drop(columns = ['score'])\n",
    "    \n",
    "    ############################################################\n",
    "    # merge aligned targets to df\n",
    "    master = dat.merge(master, on = ['clash_read'], how = 'left')\n",
    "    \n",
    "    ############################################################\n",
    "    # extend target region by 0 nt on both sides\n",
    "    records = SeqIO.to_dict(SeqIO.parse(target_fasta, 'fasta'))\n",
    "    \n",
    "    pad = 5\n",
    "    master['pad_up'] = master.apply(lambda x : x['target_start']-pad if x['target_start']-pad > 0 else 0, axis = 1 )\n",
    "    master['pad_down'] = master.apply(lambda x : x['target_end']+pad, axis = 1 )\n",
    "    master['target_pad'] = master.apply(lambda x : get_seq(records, x['target'], x['pad_up'], x['pad_down']) , axis = 1 )\n",
    "    master['ID'] = [ f'id_{i}' for i in range(0,master.shape[0]) ]\n",
    "\n",
    "    ############################################################\n",
    "    # iterate through df_merge and make RNAup input \n",
    "    lines = ''\n",
    "    for index,row in master.iterrows() : \n",
    "        lines += f\">{row['ID']}:{row['smRNA_seq']}\\n{row['target_pad']}&{row['smRNA_seq']}\\n\"\n",
    "        \n",
    "    RNAup_input = f\"{targets_aligned}.RNAup.fa\"\n",
    "    op = open(RNAup_input, 'w')\n",
    "    op.write(lines)\n",
    "    op.close() \n",
    "    lines = ''\n",
    "    \n",
    "    ############################################################\n",
    "    # RNAup folding \n",
    "    print(\"RNAup\")\n",
    "    RNAup_out = RNAup_input.replace(\".fa\", \".tsv\")\n",
    "    \n",
    "    # check size of file if lines > chunk size break into smaller files for RNA folding\n",
    "    # fasta file has two lines per entry so divide i / 2 to get number of computations\n",
    "    \n",
    "    run_RNAup(RNAup_input, outdir)\n",
    "    \n",
    "    rna_up = process_rna_up(RNAup_out)\n",
    "    refine = rna_up.query('N_piRNA_bp == 0').reset_index(drop = True)\n",
    "    rna_up = rna_up.query('N_piRNA_bp >= 4').reset_index(drop = True)\n",
    "    \n",
    "    rna_up = rna_up[['ID', 'interaction_seq', 'pairing', 'dG', 'modified_smRNA_pairing', 'target_pairing_region', 'smRNA_pairing_region']]\n",
    "    refine = refine[['ID', 'interaction_seq', 'pairing', 'dG', 'modified_smRNA_pairing', 'target_pairing_region', 'smRNA_pairing_region']]\n",
    "    ####################\n",
    "    # Do a refined search , for target RNA interactions now that potential binding sites have been identified\n",
    "    # better sequence prediction of where targeting might occur\n",
    "    \n",
    "    refine = refine.merge(master, on = ['ID'], how = 'left')\n",
    "        \n",
    "    ####################\n",
    "    # adjust target start and end from master table\n",
    "        \n",
    "    refine = refine[['ID', 'interaction_seq', 'smRNA_seq']]\n",
    "    \n",
    "    lines = ''\n",
    "    for index,row in refine.iterrows() : \n",
    "        lines += f\"\"\">{row['ID']}:{row['smRNA_seq']}\\n{ row['interaction_seq'].split(\"&\")[0].replace(\"U\", \"T\") }&{ row['smRNA_seq'] }\\n\"\"\"\n",
    "        \n",
    "    RNAup_input = f\"{targets_aligned}.RNAup.refined.fa\"\n",
    "    op = open(RNAup_input, 'w')\n",
    "    op.write(lines)\n",
    "    op.close() \n",
    "    lines = ''\n",
    "    \n",
    "    RNAup_out = RNAup_input.replace(\".fa\", \".tsv\")\n",
    "    run_RNAup(RNAup_input, outdir)\n",
    "    rna_up_refined = process_rna_up(RNAup_out)\n",
    "    rna_up_refined = rna_up_refined.query('N_piRNA_bp >= 4').reset_index(drop = True)\n",
    "    rna_up_refined = rna_up_refined[['ID', 'interaction_seq', 'pairing', 'dG', 'modified_smRNA_pairing', 'target_pairing_region', 'smRNA_pairing_region']]\n",
    "     \n",
    "    ####################\n",
    "    # merge rna_up to master\n",
    "    master_rnaup = rna_up.merge(master, on = ['ID'], how = 'left')\n",
    "    \n",
    "    ####################\n",
    "    # merge rna_up_refined to master table\n",
    "    master_rnaup_refined = rna_up_refined.merge(master, on = ['ID'], how = 'left')\n",
    "    \n",
    "    ####################\n",
    "    # merge master tables\n",
    "    master_rnaup = pd.concat([master_rnaup, master_rnaup_refined], ignore_index = True)\n",
    "    \n",
    "    master_rnaup = master_rnaup.rename(columns = {'count':'clash_count'})\n",
    "    \n",
    "    master_rnaup['clash_count'] = master_rnaup['clash_count'].astype(float)\n",
    "    \n",
    "    master_rnaup = master_rnaup.drop(columns = ['target_start', 'target_end'])\n",
    "\n",
    "    master_rnaup = master_rnaup[['ID', 'target', 'smRNA', 'clash_count', 'dG', 'modified_smRNA_pairing', 'interaction_seq', 'pairing', 'clash_read']]\n",
    "\n",
    "    ####################\n",
    "    # align target sequences to target refrence no mismatches, to get target site start and end\n",
    "    lines = ''\n",
    "    for index,row in master_rnaup.iterrows() : \n",
    "        lines += f\"\"\">{row['ID']}\\n{row['interaction_seq'].split(\"&\")[0].replace(\"U\", \"T\")}\\n\"\"\"\n",
    "        \n",
    "    rna_up_target_fa = f\"{targets_aligned}.targetRegions.fa\"\n",
    "    op = open(rna_up_target_fa, 'w')\n",
    "    op.write(lines)\n",
    "    op.close() \n",
    "    lines = ''\n",
    "    \n",
    "    bowtie1_align(rna_up_target_fa, target_index_bowtie1, rna_up_target_fa.replace(\".fa\", \"\"))\n",
    "    \n",
    "    tmp = pd.read_csv(rna_up_target_fa.replace(\".fa\", \".bed\"), \n",
    "                      sep = \"\\t\",\n",
    "                      names = ['target', 'target_start', 'target_end', 'ID'],\n",
    "                      usecols = [0,1,2,3],\n",
    "                      header = None)\n",
    "    \n",
    "    master_rnaup = master_rnaup.merge(tmp, on = ['target', 'ID'], how = 'left')\n",
    "    \n",
    "    master_rnaup = master_rnaup[['ID', 'target', 'target_start', 'target_end', 'smRNA', 'clash_count', 'dG', 'modified_smRNA_pairing', 'interaction_seq', 'pairing', 'clash_read']]\n",
    "    \n",
    "    master_rnaup['strand'] = \"-\"\n",
    "    \n",
    "    # sort values by count\n",
    "    master_rnaup = master_rnaup.sort_values(by = ['clash_count'], ascending = False).reset_index(drop = True)\n",
    "    \n",
    "    master_rnaup = master_rnaup.reset_index(drop = True)\n",
    "    print(f\"total target sites remaining == {master_rnaup.shape[0]}\")\n",
    "\n",
    "    # combine identical target sites, groupby pairing, target, smRNA, target_start, target_end, strand, \n",
    "    master_rnaup_grouped = master_rnaup.groupby( by = ['target', 'target_start', 'target_end', 'smRNA', 'modified_smRNA_pairing', 'strand', 'interaction_seq']).agg({'dG':'mean', 'clash_count':'sum'}).reset_index()\n",
    "    \n",
    "    master_rnaup_grouped = master_rnaup_grouped[['target', 'target_start', 'target_end', 'smRNA', 'clash_count', 'strand', 'dG', 'modified_smRNA_pairing', 'interaction_seq']]\n",
    "    \n",
    "    # drop na's for rnagrouped and rna\n",
    "    master_rnaup_grouped = master_rnaup_grouped.dropna().reset_index(drop = True)\n",
    "    \n",
    "    collapsed = collapse(master_rnaup_grouped)\n",
    "    \n",
    "    t2 = time.time()\n",
    "    \n",
    "    print(f\"CLASH Analysis Pipeline took {round( (t2-t1), 5)} seconds to run!\")\n",
    "    print(f\"total target sites remaining == {collapsed.shape[0]}\")\n",
    "    \n",
    "    return([ master_rnaup, master_rnaup_grouped, collapsed ])\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 382,
   "id": "e9068a50-3455-4dcf-adb1-25fe7f0c61f0",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/fs/ess/PAS1473/ben_past_projects/clash_test/test_trimmed.uni.fa\n",
      "Process smRNA alignment\n",
      "Process mRNA alignment\n",
      "RNAup\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\n",
      "The following have been reloaded with a version change:\n",
      "  1) python/3.9-2022.05 => python/3.6-conda5.2\n",
      "\n",
      "# reads processed: 9518\n",
      "# reads with at least one alignment: 9517 (99.99%)\n",
      "# reads that failed to align: 1 (0.01%)\n",
      "Reported 11476 alignments\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "total target sites remaining == 9584\n",
      "812 of 8631 target smRNA pairs processed. (9.41%)\r"
     ]
    },
    {
     "ename": "KeyboardInterrupt",
     "evalue": "",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mKeyboardInterrupt\u001b[0m                         Traceback (most recent call last)",
      "Input \u001b[0;32mIn [382]\u001b[0m, in \u001b[0;36m<cell line: 1>\u001b[0;34m()\u001b[0m\n\u001b[0;32m----> 1\u001b[0m res \u001b[38;5;241m=\u001b[39m \u001b[43mrun\u001b[49m\u001b[43m(\u001b[49m\n\u001b[1;32m      2\u001b[0m \u001b[43m    \u001b[49m\u001b[43msmRNA_fasta\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43m \u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43m/fs/ess/PCON0160/ben/genomes/c_elegans/WS279/piRNA.fa\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m,\u001b[49m\n\u001b[1;32m      3\u001b[0m \u001b[43m    \u001b[49m\u001b[43mtarget_fasta\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43m \u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43m/fs/ess/PCON0160/ben/genomes/c_elegans/WS279/ce_ws279.linc.pseudo.pc.repbase.fa\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m,\u001b[49m\n\u001b[1;32m      4\u001b[0m \u001b[43m    \u001b[49m\u001b[43mgenome_fasta\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43m \u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43m/fs/ess/PCON0160/ben/genomes/c_elegans/WS279/c_elegans.PRJNA13758.WS279.genomic.fa\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m,\u001b[49m\n\u001b[1;32m      5\u001b[0m \u001b[43m    \u001b[49m\u001b[43mfastq\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43m \u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43m/fs/ess/PAS1473/deep_sequencing/prg1CLASH/fastq/test.fq.gz\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m,\u001b[49m\n\u001b[1;32m      6\u001b[0m \u001b[43m    \u001b[49m\u001b[43mindex_dir\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43m \u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43m/fs/ess/PAS1473/ben_past_projects/clash_index\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\n\u001b[1;32m      7\u001b[0m \u001b[43m    \u001b[49m\u001b[43moutdir\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43m \u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43m/fs/ess/PAS1473/ben_past_projects/clash_test/\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m,\u001b[49m\n\u001b[1;32m      8\u001b[0m \u001b[43m    \u001b[49m\u001b[43mncores\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43m \u001b[49m\u001b[38;5;241;43m16\u001b[39;49m\n\u001b[1;32m      9\u001b[0m \u001b[43m)\u001b[49m\n",
      "Input \u001b[0;32mIn [381]\u001b[0m, in \u001b[0;36mrun\u001b[0;34m(smRNA_fasta, target_fasta, genome_fasta, fastq, index_dir, outdir, ncores)\u001b[0m\n\u001b[1;32m    253\u001b[0m \u001b[38;5;66;03m# drop na's for rnagrouped and rna\u001b[39;00m\n\u001b[1;32m    254\u001b[0m master_rnaup_grouped \u001b[38;5;241m=\u001b[39m master_rnaup_grouped\u001b[38;5;241m.\u001b[39mdropna()\u001b[38;5;241m.\u001b[39mreset_index(drop \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;01mTrue\u001b[39;00m)\n\u001b[0;32m--> 256\u001b[0m collapsed \u001b[38;5;241m=\u001b[39m \u001b[43mcollapse\u001b[49m\u001b[43m(\u001b[49m\u001b[43mmaster_rnaup_grouped\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    258\u001b[0m t2 \u001b[38;5;241m=\u001b[39m time\u001b[38;5;241m.\u001b[39mtime()\n\u001b[1;32m    260\u001b[0m \u001b[38;5;28mprint\u001b[39m(\u001b[38;5;124mf\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mCLASH Analysis Pipeline took \u001b[39m\u001b[38;5;132;01m{\u001b[39;00m\u001b[38;5;28mround\u001b[39m( (t2\u001b[38;5;241m-\u001b[39mt1), \u001b[38;5;241m5\u001b[39m)\u001b[38;5;132;01m}\u001b[39;00m\u001b[38;5;124m seconds to run!\u001b[39m\u001b[38;5;124m\"\u001b[39m)\n",
      "Input \u001b[0;32mIn [380]\u001b[0m, in \u001b[0;36mcollapse\u001b[0;34m(data)\u001b[0m\n\u001b[1;32m     46\u001b[0m k \u001b[38;5;241m+\u001b[39m\u001b[38;5;241m=\u001b[39m \u001b[38;5;241m1\u001b[39m\n\u001b[1;32m     48\u001b[0m \u001b[38;5;28mprint\u001b[39m(\u001b[38;5;124mf\u001b[39m\u001b[38;5;124m'\u001b[39m\u001b[38;5;132;01m{\u001b[39;00me\u001b[38;5;132;01m}\u001b[39;00m\u001b[38;5;124m of \u001b[39m\u001b[38;5;132;01m{\u001b[39;00m\u001b[38;5;28mlen\u001b[39m(uni)\u001b[38;5;132;01m}\u001b[39;00m\u001b[38;5;124m target smRNA pairs processed. (\u001b[39m\u001b[38;5;132;01m{\u001b[39;00m \u001b[38;5;28mround\u001b[39m(( (e)\u001b[38;5;241m/\u001b[39m(\u001b[38;5;28mlen\u001b[39m(uni)) )\u001b[38;5;241m*\u001b[39m\u001b[38;5;241m100\u001b[39m,\u001b[38;5;241m2\u001b[39m) \u001b[38;5;132;01m}\u001b[39;00m\u001b[38;5;124m%)\u001b[39m\u001b[38;5;124m'\u001b[39m, end\u001b[38;5;241m=\u001b[39m\u001b[38;5;124m'\u001b[39m\u001b[38;5;130;01m\\r\u001b[39;00m\u001b[38;5;124m'\u001b[39m)\n\u001b[0;32m---> 50\u001b[0m grped \u001b[38;5;241m=\u001b[39m \u001b[43mres\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mgroupby\u001b[49m\u001b[43m(\u001b[49m\u001b[43m[\u001b[49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[38;5;124;43mtarget\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[38;5;124;43msmRNA\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[38;5;124;43mgroup\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[43m]\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mas_index\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43m \u001b[49m\u001b[38;5;28;43;01mFalse\u001b[39;49;00m\u001b[43m)\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43magg\u001b[49m\u001b[43m(\u001b[49m\n\u001b[1;32m     51\u001b[0m \u001b[43m        \u001b[49m\u001b[43mclash_count\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43m \u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[38;5;124;43mclash_count\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[38;5;124;43msum\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[43m)\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\n\u001b[1;32m     52\u001b[0m \u001b[43m        \u001b[49m\u001b[43mdG\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43m \u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[38;5;124;43mdG\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[38;5;124;43mmin\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[43m)\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m     53\u001b[0m \u001b[43m        \u001b[49m\u001b[43mstart\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43m \u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[38;5;124;43mtarget_start\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[38;5;124;43mmin\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[43m)\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\n\u001b[1;32m     54\u001b[0m \u001b[43m        \u001b[49m\u001b[43mend\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43m \u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[38;5;124;43mtarget_end\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[38;5;124;43mmax\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[43m)\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m     55\u001b[0m \u001b[43m        \u001b[49m\u001b[43mpairing\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43m \u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[38;5;124;43mmodified_smRNA_pairing\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[38;5;124;43munique\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[43m)\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m     56\u001b[0m \u001b[43m        \u001b[49m\u001b[43minteraction\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43m \u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[38;5;124;43minteraction_seq\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[38;5;124;43munique\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[43m)\u001b[49m\u001b[43m    \u001b[49m\n\u001b[1;32m     57\u001b[0m \u001b[43m    \u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m     58\u001b[0m grped[\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mstrand\u001b[39m\u001b[38;5;124m'\u001b[39m] \u001b[38;5;241m=\u001b[39m \u001b[38;5;124m'\u001b[39m\u001b[38;5;124m+\u001b[39m\u001b[38;5;124m'\u001b[39m\n\u001b[1;32m     59\u001b[0m reformat \u001b[38;5;241m=\u001b[39m grped[[\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mtarget\u001b[39m\u001b[38;5;124m'\u001b[39m, \u001b[38;5;124m'\u001b[39m\u001b[38;5;124mstart\u001b[39m\u001b[38;5;124m'\u001b[39m, \u001b[38;5;124m'\u001b[39m\u001b[38;5;124mend\u001b[39m\u001b[38;5;124m'\u001b[39m, \u001b[38;5;124m'\u001b[39m\u001b[38;5;124msmRNA\u001b[39m\u001b[38;5;124m'\u001b[39m, \u001b[38;5;124m'\u001b[39m\u001b[38;5;124mclash_count\u001b[39m\u001b[38;5;124m'\u001b[39m, \u001b[38;5;124m'\u001b[39m\u001b[38;5;124mstrand\u001b[39m\u001b[38;5;124m'\u001b[39m, \u001b[38;5;124m'\u001b[39m\u001b[38;5;124mdG\u001b[39m\u001b[38;5;124m'\u001b[39m, \u001b[38;5;124m'\u001b[39m\u001b[38;5;124minteraction\u001b[39m\u001b[38;5;124m'\u001b[39m, \u001b[38;5;124m'\u001b[39m\u001b[38;5;124mpairing\u001b[39m\u001b[38;5;124m'\u001b[39m]]\n",
      "File \u001b[0;32m/apps/python/3.9-2022.05/lib/python3.9/site-packages/pandas/core/groupby/generic.py:869\u001b[0m, in \u001b[0;36mDataFrameGroupBy.aggregate\u001b[0;34m(self, func, engine, engine_kwargs, *args, **kwargs)\u001b[0m\n\u001b[1;32m    866\u001b[0m func \u001b[38;5;241m=\u001b[39m maybe_mangle_lambdas(func)\n\u001b[1;32m    868\u001b[0m op \u001b[38;5;241m=\u001b[39m GroupByApply(\u001b[38;5;28mself\u001b[39m, func, args, kwargs)\n\u001b[0;32m--> 869\u001b[0m result \u001b[38;5;241m=\u001b[39m \u001b[43mop\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43magg\u001b[49m\u001b[43m(\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    870\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m is_dict_like(func) \u001b[38;5;129;01mand\u001b[39;00m result \u001b[38;5;129;01mis\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m \u001b[38;5;28;01mNone\u001b[39;00m:\n\u001b[1;32m    871\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m result\n",
      "File \u001b[0;32m/apps/python/3.9-2022.05/lib/python3.9/site-packages/pandas/core/apply.py:168\u001b[0m, in \u001b[0;36mApply.agg\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m    165\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mapply_str()\n\u001b[1;32m    167\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m is_dict_like(arg):\n\u001b[0;32m--> 168\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43magg_dict_like\u001b[49m\u001b[43m(\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    169\u001b[0m \u001b[38;5;28;01melif\u001b[39;00m is_list_like(arg):\n\u001b[1;32m    170\u001b[0m     \u001b[38;5;66;03m# we require a list, but not a 'str'\u001b[39;00m\n\u001b[1;32m    171\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39magg_list_like()\n",
      "File \u001b[0;32m/apps/python/3.9-2022.05/lib/python3.9/site-packages/pandas/core/apply.py:475\u001b[0m, in \u001b[0;36mApply.agg_dict_like\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m    472\u001b[0m     results \u001b[38;5;241m=\u001b[39m {key: colg\u001b[38;5;241m.\u001b[39magg(how) \u001b[38;5;28;01mfor\u001b[39;00m key, how \u001b[38;5;129;01min\u001b[39;00m arg\u001b[38;5;241m.\u001b[39mitems()}\n\u001b[1;32m    473\u001b[0m \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[1;32m    474\u001b[0m     \u001b[38;5;66;03m# key used for column selection and output\u001b[39;00m\n\u001b[0;32m--> 475\u001b[0m     results \u001b[38;5;241m=\u001b[39m {\n\u001b[1;32m    476\u001b[0m         key: obj\u001b[38;5;241m.\u001b[39m_gotitem(key, ndim\u001b[38;5;241m=\u001b[39m\u001b[38;5;241m1\u001b[39m)\u001b[38;5;241m.\u001b[39magg(how) \u001b[38;5;28;01mfor\u001b[39;00m key, how \u001b[38;5;129;01min\u001b[39;00m arg\u001b[38;5;241m.\u001b[39mitems()\n\u001b[1;32m    477\u001b[0m     }\n\u001b[1;32m    479\u001b[0m \u001b[38;5;66;03m# set the final keys\u001b[39;00m\n\u001b[1;32m    480\u001b[0m keys \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mlist\u001b[39m(arg\u001b[38;5;241m.\u001b[39mkeys())\n",
      "File \u001b[0;32m/apps/python/3.9-2022.05/lib/python3.9/site-packages/pandas/core/apply.py:476\u001b[0m, in \u001b[0;36m<dictcomp>\u001b[0;34m(.0)\u001b[0m\n\u001b[1;32m    472\u001b[0m     results \u001b[38;5;241m=\u001b[39m {key: colg\u001b[38;5;241m.\u001b[39magg(how) \u001b[38;5;28;01mfor\u001b[39;00m key, how \u001b[38;5;129;01min\u001b[39;00m arg\u001b[38;5;241m.\u001b[39mitems()}\n\u001b[1;32m    473\u001b[0m \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[1;32m    474\u001b[0m     \u001b[38;5;66;03m# key used for column selection and output\u001b[39;00m\n\u001b[1;32m    475\u001b[0m     results \u001b[38;5;241m=\u001b[39m {\n\u001b[0;32m--> 476\u001b[0m         key: \u001b[43mobj\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_gotitem\u001b[49m\u001b[43m(\u001b[49m\u001b[43mkey\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mndim\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;241;43m1\u001b[39;49m\u001b[43m)\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43magg\u001b[49m\u001b[43m(\u001b[49m\u001b[43mhow\u001b[49m\u001b[43m)\u001b[49m \u001b[38;5;28;01mfor\u001b[39;00m key, how \u001b[38;5;129;01min\u001b[39;00m arg\u001b[38;5;241m.\u001b[39mitems()\n\u001b[1;32m    477\u001b[0m     }\n\u001b[1;32m    479\u001b[0m \u001b[38;5;66;03m# set the final keys\u001b[39;00m\n\u001b[1;32m    480\u001b[0m keys \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mlist\u001b[39m(arg\u001b[38;5;241m.\u001b[39mkeys())\n",
      "File \u001b[0;32m/apps/python/3.9-2022.05/lib/python3.9/site-packages/pandas/core/groupby/generic.py:271\u001b[0m, in \u001b[0;36mSeriesGroupBy.aggregate\u001b[0;34m(self, func, engine, engine_kwargs, *args, **kwargs)\u001b[0m\n\u001b[1;32m    267\u001b[0m \u001b[38;5;28;01melif\u001b[39;00m \u001b[38;5;28misinstance\u001b[39m(func, abc\u001b[38;5;241m.\u001b[39mIterable):\n\u001b[1;32m    268\u001b[0m     \u001b[38;5;66;03m# Catch instances of lists / tuples\u001b[39;00m\n\u001b[1;32m    269\u001b[0m     \u001b[38;5;66;03m# but not the class list / tuple itself.\u001b[39;00m\n\u001b[1;32m    270\u001b[0m     func \u001b[38;5;241m=\u001b[39m maybe_mangle_lambdas(func)\n\u001b[0;32m--> 271\u001b[0m     ret \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_aggregate_multiple_funcs\u001b[49m\u001b[43m(\u001b[49m\u001b[43mfunc\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    272\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m relabeling:\n\u001b[1;32m    273\u001b[0m         \u001b[38;5;66;03m# error: Incompatible types in assignment (expression has type\u001b[39;00m\n\u001b[1;32m    274\u001b[0m         \u001b[38;5;66;03m# \"Optional[List[str]]\", variable has type \"Index\")\u001b[39;00m\n\u001b[1;32m    275\u001b[0m         ret\u001b[38;5;241m.\u001b[39mcolumns \u001b[38;5;241m=\u001b[39m columns  \u001b[38;5;66;03m# type: ignore[assignment]\u001b[39;00m\n",
      "File \u001b[0;32m/apps/python/3.9-2022.05/lib/python3.9/site-packages/pandas/core/groupby/generic.py:326\u001b[0m, in \u001b[0;36mSeriesGroupBy._aggregate_multiple_funcs\u001b[0;34m(self, arg)\u001b[0m\n\u001b[1;32m    323\u001b[0m \u001b[38;5;28;01mfor\u001b[39;00m idx, (name, func) \u001b[38;5;129;01min\u001b[39;00m \u001b[38;5;28menumerate\u001b[39m(arg):\n\u001b[1;32m    325\u001b[0m     key \u001b[38;5;241m=\u001b[39m base\u001b[38;5;241m.\u001b[39mOutputKey(label\u001b[38;5;241m=\u001b[39mname, position\u001b[38;5;241m=\u001b[39midx)\n\u001b[0;32m--> 326\u001b[0m     results[key] \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43maggregate\u001b[49m\u001b[43m(\u001b[49m\u001b[43mfunc\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    328\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28many\u001b[39m(\u001b[38;5;28misinstance\u001b[39m(x, DataFrame) \u001b[38;5;28;01mfor\u001b[39;00m x \u001b[38;5;129;01min\u001b[39;00m results\u001b[38;5;241m.\u001b[39mvalues()):\n\u001b[1;32m    329\u001b[0m     \u001b[38;5;28;01mfrom\u001b[39;00m \u001b[38;5;21;01mpandas\u001b[39;00m \u001b[38;5;28;01mimport\u001b[39;00m concat\n",
      "File \u001b[0;32m/apps/python/3.9-2022.05/lib/python3.9/site-packages/pandas/core/groupby/generic.py:265\u001b[0m, in \u001b[0;36mSeriesGroupBy.aggregate\u001b[0;34m(self, func, engine, engine_kwargs, *args, **kwargs)\u001b[0m\n\u001b[1;32m    262\u001b[0m     kwargs \u001b[38;5;241m=\u001b[39m {}\n\u001b[1;32m    264\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28misinstance\u001b[39m(func, \u001b[38;5;28mstr\u001b[39m):\n\u001b[0;32m--> 265\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28;43mgetattr\u001b[39;49m\u001b[43m(\u001b[49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mfunc\u001b[49m\u001b[43m)\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43margs\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43mkwargs\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    267\u001b[0m \u001b[38;5;28;01melif\u001b[39;00m \u001b[38;5;28misinstance\u001b[39m(func, abc\u001b[38;5;241m.\u001b[39mIterable):\n\u001b[1;32m    268\u001b[0m     \u001b[38;5;66;03m# Catch instances of lists / tuples\u001b[39;00m\n\u001b[1;32m    269\u001b[0m     \u001b[38;5;66;03m# but not the class list / tuple itself.\u001b[39;00m\n\u001b[1;32m    270\u001b[0m     func \u001b[38;5;241m=\u001b[39m maybe_mangle_lambdas(func)\n",
      "File \u001b[0;32m/apps/python/3.9-2022.05/lib/python3.9/site-packages/pandas/core/groupby/groupby.py:950\u001b[0m, in \u001b[0;36mGroupBy._make_wrapper.<locals>.wrapper\u001b[0;34m(*args, **kwargs)\u001b[0m\n\u001b[1;32m    947\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m name \u001b[38;5;129;01min\u001b[39;00m base\u001b[38;5;241m.\u001b[39mplotting_methods:\n\u001b[1;32m    948\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mapply(curried)\n\u001b[0;32m--> 950\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_python_apply_general\u001b[49m\u001b[43m(\u001b[49m\u001b[43mcurried\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_obj_with_exclusions\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m/apps/python/3.9-2022.05/lib/python3.9/site-packages/pandas/core/groupby/groupby.py:1455\u001b[0m, in \u001b[0;36mGroupBy._python_apply_general\u001b[0;34m(self, f, data, not_indexed_same)\u001b[0m\n\u001b[1;32m   1429\u001b[0m \u001b[38;5;129m@final\u001b[39m\n\u001b[1;32m   1430\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m \u001b[38;5;21m_python_apply_general\u001b[39m(\n\u001b[1;32m   1431\u001b[0m     \u001b[38;5;28mself\u001b[39m,\n\u001b[0;32m   (...)\u001b[0m\n\u001b[1;32m   1434\u001b[0m     not_indexed_same: \u001b[38;5;28mbool\u001b[39m \u001b[38;5;241m|\u001b[39m \u001b[38;5;28;01mNone\u001b[39;00m \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;01mNone\u001b[39;00m,\n\u001b[1;32m   1435\u001b[0m ) \u001b[38;5;241m-\u001b[39m\u001b[38;5;241m>\u001b[39m DataFrame \u001b[38;5;241m|\u001b[39m Series:\n\u001b[1;32m   1436\u001b[0m     \u001b[38;5;124;03m\"\"\"\u001b[39;00m\n\u001b[1;32m   1437\u001b[0m \u001b[38;5;124;03m    Apply function f in python space\u001b[39;00m\n\u001b[1;32m   1438\u001b[0m \n\u001b[0;32m   (...)\u001b[0m\n\u001b[1;32m   1453\u001b[0m \u001b[38;5;124;03m        data after applying f\u001b[39;00m\n\u001b[1;32m   1454\u001b[0m \u001b[38;5;124;03m    \"\"\"\u001b[39;00m\n\u001b[0;32m-> 1455\u001b[0m     values, mutated \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mgrouper\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mapply\u001b[49m\u001b[43m(\u001b[49m\u001b[43mf\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mdata\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43maxis\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m   1457\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m not_indexed_same \u001b[38;5;129;01mis\u001b[39;00m \u001b[38;5;28;01mNone\u001b[39;00m:\n\u001b[1;32m   1458\u001b[0m         not_indexed_same \u001b[38;5;241m=\u001b[39m mutated \u001b[38;5;129;01mor\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mmutated\n",
      "File \u001b[0;32m/apps/python/3.9-2022.05/lib/python3.9/site-packages/pandas/core/groupby/ops.py:761\u001b[0m, in \u001b[0;36mBaseGrouper.apply\u001b[0;34m(self, f, data, axis)\u001b[0m\n\u001b[1;32m    759\u001b[0m \u001b[38;5;66;03m# group might be modified\u001b[39;00m\n\u001b[1;32m    760\u001b[0m group_axes \u001b[38;5;241m=\u001b[39m group\u001b[38;5;241m.\u001b[39maxes\n\u001b[0;32m--> 761\u001b[0m res \u001b[38;5;241m=\u001b[39m \u001b[43mf\u001b[49m\u001b[43m(\u001b[49m\u001b[43mgroup\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    762\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m mutated \u001b[38;5;129;01mand\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m _is_indexed_like(res, group_axes, axis):\n\u001b[1;32m    763\u001b[0m     mutated \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;01mTrue\u001b[39;00m\n",
      "File \u001b[0;32m/apps/python/3.9-2022.05/lib/python3.9/site-packages/pandas/core/groupby/groupby.py:939\u001b[0m, in \u001b[0;36mGroupBy._make_wrapper.<locals>.wrapper.<locals>.curried\u001b[0;34m(x)\u001b[0m\n\u001b[1;32m    938\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m \u001b[38;5;21mcurried\u001b[39m(x):\n\u001b[0;32m--> 939\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[43mf\u001b[49m\u001b[43m(\u001b[49m\u001b[43mx\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43margs\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43mkwargs\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m/apps/python/3.9-2022.05/lib/python3.9/site-packages/pandas/core/series.py:2088\u001b[0m, in \u001b[0;36mSeries.unique\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m   2030\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m \u001b[38;5;21munique\u001b[39m(\u001b[38;5;28mself\u001b[39m) \u001b[38;5;241m-\u001b[39m\u001b[38;5;241m>\u001b[39m ArrayLike:\n\u001b[1;32m   2031\u001b[0m     \u001b[38;5;124;03m\"\"\"\u001b[39;00m\n\u001b[1;32m   2032\u001b[0m \u001b[38;5;124;03m    Return unique values of Series object.\u001b[39;00m\n\u001b[1;32m   2033\u001b[0m \n\u001b[0;32m   (...)\u001b[0m\n\u001b[1;32m   2086\u001b[0m \u001b[38;5;124;03m    Categories (3, object): ['a' < 'b' < 'c']\u001b[39;00m\n\u001b[1;32m   2087\u001b[0m \u001b[38;5;124;03m    \"\"\"\u001b[39;00m\n\u001b[0;32m-> 2088\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28;43msuper\u001b[39;49m\u001b[43m(\u001b[49m\u001b[43m)\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43munique\u001b[49m\u001b[43m(\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m/apps/python/3.9-2022.05/lib/python3.9/site-packages/pandas/core/base.py:989\u001b[0m, in \u001b[0;36mIndexOpsMixin.unique\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m    987\u001b[0m             result \u001b[38;5;241m=\u001b[39m np\u001b[38;5;241m.\u001b[39masarray(result)\n\u001b[1;32m    988\u001b[0m \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[0;32m--> 989\u001b[0m     result \u001b[38;5;241m=\u001b[39m \u001b[43munique1d\u001b[49m\u001b[43m(\u001b[49m\u001b[43mvalues\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    991\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m result\n",
      "File \u001b[0;32m/apps/python/3.9-2022.05/lib/python3.9/site-packages/pandas/core/algorithms.py:432\u001b[0m, in \u001b[0;36munique\u001b[0;34m(values)\u001b[0m\n\u001b[1;32m    337\u001b[0m \u001b[38;5;124;03m\"\"\"\u001b[39;00m\n\u001b[1;32m    338\u001b[0m \u001b[38;5;124;03mReturn unique values based on a hash table.\u001b[39;00m\n\u001b[1;32m    339\u001b[0m \n\u001b[0;32m   (...)\u001b[0m\n\u001b[1;32m    428\u001b[0m \u001b[38;5;124;03marray([('a', 'b'), ('b', 'a'), ('a', 'c')], dtype=object)\u001b[39;00m\n\u001b[1;32m    429\u001b[0m \u001b[38;5;124;03m\"\"\"\u001b[39;00m\n\u001b[1;32m    430\u001b[0m values \u001b[38;5;241m=\u001b[39m _ensure_arraylike(values)\n\u001b[0;32m--> 432\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m \u001b[43mis_extension_array_dtype\u001b[49m\u001b[43m(\u001b[49m\u001b[43mvalues\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mdtype\u001b[49m\u001b[43m)\u001b[49m:\n\u001b[1;32m    433\u001b[0m     \u001b[38;5;66;03m# Dispatch to extension dtype's unique.\u001b[39;00m\n\u001b[1;32m    434\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m values\u001b[38;5;241m.\u001b[39munique()\n\u001b[1;32m    436\u001b[0m original \u001b[38;5;241m=\u001b[39m values\n",
      "File \u001b[0;32m/apps/python/3.9-2022.05/lib/python3.9/site-packages/pandas/core/dtypes/common.py:1475\u001b[0m, in \u001b[0;36mis_extension_array_dtype\u001b[0;34m(arr_or_dtype)\u001b[0m\n\u001b[1;32m   1430\u001b[0m \u001b[38;5;124;03m\"\"\"\u001b[39;00m\n\u001b[1;32m   1431\u001b[0m \u001b[38;5;124;03mCheck if an object is a pandas extension array type.\u001b[39;00m\n\u001b[1;32m   1432\u001b[0m \n\u001b[0;32m   (...)\u001b[0m\n\u001b[1;32m   1472\u001b[0m \u001b[38;5;124;03mFalse\u001b[39;00m\n\u001b[1;32m   1473\u001b[0m \u001b[38;5;124;03m\"\"\"\u001b[39;00m\n\u001b[1;32m   1474\u001b[0m dtype \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mgetattr\u001b[39m(arr_or_dtype, \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mdtype\u001b[39m\u001b[38;5;124m\"\u001b[39m, arr_or_dtype)\n\u001b[0;32m-> 1475\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28;43misinstance\u001b[39;49m\u001b[43m(\u001b[49m\u001b[43mdtype\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mExtensionDtype\u001b[49m\u001b[43m)\u001b[49m:\n\u001b[1;32m   1476\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28;01mTrue\u001b[39;00m\n\u001b[1;32m   1477\u001b[0m \u001b[38;5;28;01melif\u001b[39;00m \u001b[38;5;28misinstance\u001b[39m(dtype, np\u001b[38;5;241m.\u001b[39mdtype):\n",
      "\u001b[0;31mKeyboardInterrupt\u001b[0m: "
     ]
    }
   ],
   "source": [
    "res = run(\n",
    "    smRNA_fasta = \"/fs/ess/PCON0160/ben/genomes/c_elegans/WS279/piRNA.fa\",\n",
    "    target_fasta = \"/fs/ess/PCON0160/ben/genomes/c_elegans/WS279/ce_ws279.linc.pseudo.pc.repbase.fa\",\n",
    "    genome_fasta = \"/fs/ess/PCON0160/ben/genomes/c_elegans/WS279/c_elegans.PRJNA13758.WS279.genomic.fa\",\n",
    "    fastq = \"/fs/ess/PAS1473/deep_sequencing/prg1CLASH/fastq/test.fq.gz\",\n",
    "    index_dir = \"/fs/ess/PAS1473/ben_past_projects/clash_index\", \n",
    "    outdir = \"/fs/ess/PAS1473/ben_past_projects/clash_test/\",\n",
    "    ncores = 16\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3cc9453c-17e3-44fd-af0e-0a376443b8b0",
   "metadata": {},
   "outputs": [],
   "source": [
    "raw = res[0]\n",
    "grouped = res[1]\n",
    "collapsed = res[2]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fa141ddd-3cec-4f15-b405-fc922b2de094",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3.9 (Conda 2022.05) [python/3.9-2022.05]",
   "language": "python",
   "name": "python39_202205"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
